{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "801bd2cd",
   "metadata": {},
   "source": [
    "# Hyperparameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f174ad6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/ESMSec/data/cell_cycle/dataset-cell-cycle_2:3.csv\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.special import softmax\n",
    "from sklearn.model_selection import train_test_split\n",
    "import random\n",
    "import torch\n",
    "\n",
    "import utils.dataset_functions as dataf\n",
    "\n",
    "# Initializations\n",
    "SEED=42\n",
    "random.seed(SEED)\n",
    "np.random.seed(SEED)\n",
    "\n",
    "# Directory containing MSigDB JSON files\n",
    "JSON_DIR = \"/home/gdallagl/myworkdir/data/MSigDB/msigdb_v2025.1.Hs_json_files_to_download_locally\"\n",
    "\n",
    "# cellcycel geensets saving apth\n",
    "CELL_CYCLE_CSV_PATH = \"/home/gdallagl/myworkdir/data/MSigDB/cell_cycle_genesets.csv\"\n",
    "\n",
    "# Guaranted genes list\n",
    "GUARANTEED_GENES_PATH = \"/home/gdallagl/myworkdir/data/MSigDB/julies_cycling_signatures_cancer.tsv\"\n",
    "\n",
    "# Updated keywords pattern with word boundaries to avoid false matches\n",
    "KEYWORDS_PATTERN = \"|\".join([\n",
    "    \"_PROLIFERATION_\", #avoid \"proliferative\" \"proliferator\"\n",
    "    \"_CYCLING\",  # avoid \"recycling\"\n",
    "    \"CELL_CYCLE\",\n",
    "    \"_CC_\", \"_G1_\", \"_S_PHASE_\", \"_G2_\", \"_M_PHASE_\", # avoid \"aCCumbens\"\n",
    "    \"MITOSIS\", \"MITOTIC\",\n",
    "    \"CDK\",\n",
    "    \"CHECKPOINT\"\n",
    "])\n",
    "\n",
    "# Exclusion pattern\n",
    "EXCLUSION_PATTERN = r\"MEIOTIC|MEIOSIS|FATTY_ACID_CYCLING_MODEL\"\n",
    "\n",
    "\n",
    "# KEYWORDS_PATTERN = \"|\".join([\n",
    "#     # Key regulatory phrases\n",
    "#     \"_CELL_CYCLE_CHECKPOINT\",\n",
    "#     \"CELL_CYCLE_REGULATION\",\n",
    "#     \"CELL_CYCLE_CONTROL\",\n",
    "\n",
    "#     \"HALLMARK_E2F_TARGETS\", \n",
    "#     \" HALLMARK_G2M_CHECKPOINT\",\n",
    "#     # Core regulators\n",
    "#     \"CDK\", \"_CYCLIN\",\n",
    "\n",
    "#     # Transitions\n",
    "#     \"_G1_S\", \"G1_S_\", \"G2_M\", \"_S_PHASE_\", \"M_PHASE\", \"MITOTIC_CHECKPOINT\"\n",
    "# ])\n",
    "# # Exclusion pattern\n",
    "# EXCLUSION_PATTERN = r\"MEIOTIC|MEIOSIS|FATTY_ACID_CYCLING_MODEL|CDK5|CANCER|APOPTOSIS|RECYCLING|SPINDLE|KINETOCHORE|CYTOKINESIS|CENTROSOME|KEGG_MEDICUS_PATHOGEN|KEGG_MEDICUS_VARIANT|KEGG_MEDICUS_REFERENCE|PREDICTED|HE_LIM_SUN_FETAL_LUNG|MALIGNANT|MORF_CDK2\"\n",
    "\n",
    "# Human proteome path\n",
    "HUMAN_PROTEOME_PATH = \"/home/gdallagl/myworkdir/data/UniRef50/human_proteome.tsv\"\n",
    "\n",
    "# mapping protein-gene apth\n",
    "MAPPING_PATH = \"/home/gdallagl/myworkdir/ESMSec/data/UniRef50/HUMAN_9606_idmapping.dat\"\n",
    "\n",
    "# Minimum frequency threshold for filtering ambiguous genes\n",
    "MIN_FREQ_AMBIGOUS = 2\n",
    "\n",
    "# min number of postive samples per positive cluster\n",
    "MIN_SAMPLE_N_POSITIVE = 2\n",
    "\n",
    "# multiplocative factor to kwno how much morenegative class to sample\n",
    "NEGATIVE_CLASS_MULT = 3\n",
    "\n",
    "# if use as postive class only the guaranted genes\n",
    "ONLY_GUARANTEED = False\n",
    "only_guaranteed = \"only-guaranteed_\" if ONLY_GUARANTEED else \"\"\n",
    "\n",
    "# id samplepostive class or take all postive egens\n",
    "TAKE_ALL_POSITIVE_GENES = False\n",
    "\n",
    "# If use only curated (swissprot) proteins or not\n",
    "ONLY_REVIEWED_PROTEINS = True\n",
    "\n",
    "# savifn csv datset\n",
    "FINAL_DATASET_PATH = f\"/home/gdallagl/myworkdir/ESMSec/data/cell_cycle/dataset-cell-cycle_{only_guaranteed}{MIN_SAMPLE_N_POSITIVE}:{NEGATIVE_CLASS_MULT}.csv\"\n",
    "print(FINAL_DATASET_PATH)\n",
    "\n",
    "# Autorelaod\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f066315c",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_names = [\n",
    "    'Gene Name', 'AC', 'Variant AA Change', 'Source DB ID',\n",
    "    'Consequence Type', 'Clinical Significance', 'Phenotype/Disease',\n",
    "    'Phenotype/Disease Source', 'Cytogenetic Band', 'Chromosome Coordinate',\n",
    "    'Ensembl gene ID', 'Ensembl transcript ID', 'Ensembl translation ID',\n",
    "    'Evidence'\n",
    "]\n",
    "df_original = pd.read_csv(\"/home/gdallagl/myworkdir/ESMSec/data/mutations/uniprot_homo_sapiens_variation.txt\", skiprows=164, sep=\"\\t\", #nrows=98,\n",
    "            header=None,           # Crucial: Tells pandas there is NO header in the lines being read\n",
    "            names=column_names,     # Assign the custom column names\n",
    "            usecols=['Gene Name', 'Variant AA Change', 'Source DB ID', 'Consequence Type', 'Clinical Significance', 'Phenotype/Disease',] # **Only load these two columns**\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5a5eb7c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Variant AA Change</th>\n",
       "      <th>Source DB ID</th>\n",
       "      <th>Consequence Type</th>\n",
       "      <th>Clinical Significance</th>\n",
       "      <th>Phenotype/Disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NSRP1</td>\n",
       "      <td>p.Lys30Glu</td>\n",
       "      <td>RCV004545615</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>NSRP1-related disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>NSRP1</td>\n",
       "      <td>p.Asp273Glu</td>\n",
       "      <td>RCV003493117</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance</td>\n",
       "      <td>Neurodevelopmental disorder with spasticity, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027</th>\n",
       "      <td>SERPINB6</td>\n",
       "      <td>p.Met5Ile</td>\n",
       "      <td>RCV001783729</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>Autosomal recessive nonsyndromic hearing loss 91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1035</th>\n",
       "      <td>SERPINB6</td>\n",
       "      <td>p.Val45Ile</td>\n",
       "      <td>RCV004757137</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance, Likely benign</td>\n",
       "      <td>SERPINB6-related disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1036</th>\n",
       "      <td>SERPINB6</td>\n",
       "      <td>p.Val45Ile</td>\n",
       "      <td>RCV001335154</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance, Likely benign</td>\n",
       "      <td>Autosomal recessive nonsyndromic hearing loss 91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687810</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asp229Val</td>\n",
       "      <td>RCV001001047</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687819</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Val233Leu</td>\n",
       "      <td>RCV003486174</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687861</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Leu270Pro</td>\n",
       "      <td>RCV003133084</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687866</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Gly275Arg</td>\n",
       "      <td>RCV000004745</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687900</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Glu301Asp</td>\n",
       "      <td>RCV003134568</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3669831 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene Name Variant AA Change  Source DB ID  Consequence Type  \\\n",
       "33           NSRP1        p.Lys30Glu  RCV004545615  missense variant   \n",
       "273          NSRP1       p.Asp273Glu  RCV003493117  missense variant   \n",
       "1027      SERPINB6         p.Met5Ile  RCV001783729  missense variant   \n",
       "1035      SERPINB6        p.Val45Ile  RCV004757137  missense variant   \n",
       "1036      SERPINB6        p.Val45Ile  RCV001335154  missense variant   \n",
       "...            ...               ...           ...               ...   \n",
       "35687810    NT5C3A       p.Asp229Val  RCV001001047  missense variant   \n",
       "35687819    NT5C3A       p.Val233Leu  RCV003486174  missense variant   \n",
       "35687861    NT5C3A       p.Leu270Pro  RCV003133084  missense variant   \n",
       "35687866    NT5C3A       p.Gly275Arg  RCV000004745  missense variant   \n",
       "35687900    NT5C3A       p.Glu301Asp  RCV003134568  missense variant   \n",
       "\n",
       "                                     Clinical Significance  \\\n",
       "33                                           Likely benign   \n",
       "273                      Variant of uncertain significance   \n",
       "1027                                     Likely pathogenic   \n",
       "1035      Variant of uncertain significance, Likely benign   \n",
       "1036      Variant of uncertain significance, Likely benign   \n",
       "...                                                    ...   \n",
       "35687810                 Variant of uncertain significance   \n",
       "35687819                 Variant of uncertain significance   \n",
       "35687861                 Variant of uncertain significance   \n",
       "35687866                                        Pathogenic   \n",
       "35687900                 Variant of uncertain significance   \n",
       "\n",
       "                                          Phenotype/Disease  \n",
       "33                                   NSRP1-related disorder  \n",
       "273       Neurodevelopmental disorder with spasticity, s...  \n",
       "1027       Autosomal recessive nonsyndromic hearing loss 91  \n",
       "1035                              SERPINB6-related disorder  \n",
       "1036       Autosomal recessive nonsyndromic hearing loss 91  \n",
       "...                                                     ...  \n",
       "35687810  Hemolytic anemia due to pyrimidine 5' nucleoti...  \n",
       "35687819  Hemolytic anemia due to pyrimidine 5' nucleoti...  \n",
       "35687861  Hemolytic anemia due to pyrimidine 5' nucleoti...  \n",
       "35687866  Hemolytic anemia due to pyrimidine 5' nucleoti...  \n",
       "35687900  Hemolytic anemia due to pyrimidine 5' nucleoti...  \n",
       "\n",
       "[3669831 rows x 6 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df = df_original[\n",
    "    (df_original[\"Clinical Significance\"] != \"-\") &\n",
    "    (~ df_original[\"Clinical Significance\"].isna()) &\n",
    "    (df_original[\"Consequence Type\"].isin(['missense variant']))\n",
    "].copy()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "7242e225",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  Clinical Significance  \\\n",
      "33                                        Likely benign   \n",
      "273                   Variant of uncertain significance   \n",
      "1027                                  Likely pathogenic   \n",
      "1035   Variant of uncertain significance, Likely benign   \n",
      "1036   Variant of uncertain significance, Likely benign   \n",
      "1046                  Variant of uncertain significance   \n",
      "1052                  Variant of uncertain significance   \n",
      "1057                                             Benign   \n",
      "1059   Likely benign, Variant of uncertain significance   \n",
      "1073                                             Benign   \n",
      "1111  Likely pathogenic, Variant of uncertain signif...   \n",
      "1118                  Variant of uncertain significance   \n",
      "\n",
      "     Simplified Significance  \n",
      "33                    Benign  \n",
      "273             Unclassified  \n",
      "1027              Pathogenic  \n",
      "1035                  Benign  \n",
      "1036                  Benign  \n",
      "1046            Unclassified  \n",
      "1052            Unclassified  \n",
      "1057                  Benign  \n",
      "1059                  Benign  \n",
      "1073                  Benign  \n",
      "1111              Pathogenic  \n",
      "1118            Unclassified  \n",
      "\n",
      "New Categories Count:\n",
      "Simplified Significance\n",
      "Unclassified             2261014\n",
      "Benign                    869919\n",
      "Pathogenic                477295\n",
      "Conflicting/Ambiguous      60991\n",
      "Other/Modifier               612\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "PATHOGENIC_TERMS = ['Pathogenic', 'Likely pathogenic']\n",
    "BENIGN_TERMS = ['Benign', 'Likely benign']\n",
    "VUS_TERM = 'Variant of uncertain significance'\n",
    "OTHER_TERMS = ['Risk factor', 'Drug response', 'Protective', 'association', 'Conflicting interpretations of pathogenicity']\n",
    "\n",
    "\n",
    "def simplify_clinical_significance(text):\n",
    "    \"\"\"Refactors a clinical significance string into a smaller category.\"\"\"\n",
    "    if not isinstance(text, str):\n",
    "        return 'Unknown'\n",
    "    \n",
    "    # Standardize the string: lowercase and split by comma/space to get individual terms\n",
    "    terms = set([term.strip() for part in text.lower().split(',') for term in part.split() if term.strip()])\n",
    "\n",
    "    # Check for core classifications\n",
    "    is_pathogenic = any(p_term.lower() in terms for p_term in PATHOGENIC_TERMS)\n",
    "    is_benign = any(b_term.lower() in terms for b_term in BENIGN_TERMS)\n",
    "    is_vus = VUS_TERM.lower() in terms\n",
    "    \n",
    "    # Check for other modifiers\n",
    "    is_other = any(o_term.lower().replace(' ', '_') in terms for o_term in OTHER_TERMS)\n",
    "\n",
    "    # ------------------------------------------------------------------\n",
    "    # Apply Classification Logic\n",
    "    # ------------------------------------------------------------------\n",
    "    \n",
    "    if is_pathogenic and is_benign:\n",
    "        # Mixed categories (e.g., Pathogenic, Benign) are highly ambiguous and should be flagged\n",
    "        return 'Conflicting/Ambiguous'\n",
    "    \n",
    "    elif is_pathogenic:\n",
    "        return 'Pathogenic'\n",
    "    \n",
    "    elif is_benign:\n",
    "        return 'Benign'\n",
    "    \n",
    "    elif is_vus:\n",
    "        # Check if VUS is combined with only \"other\" terms, but not P/LP/B/LB\n",
    "        return 'VUS'\n",
    "        \n",
    "    elif is_other:\n",
    "        # Handle single modifiers that didn't contain P/LP/B/LB/VUS\n",
    "        return 'Other/Modifier'\n",
    "\n",
    "    return 'Unclassified'\n",
    "\n",
    "df['Simplified Significance'] = df['Clinical Significance'].apply(simplify_clinical_significance)\n",
    "\n",
    "# Display the new categories\n",
    "print(df[['Clinical Significance', 'Simplified Significance']].head(12))\n",
    "print(\"\\nNew Categories Count:\")\n",
    "print(df['Simplified Significance'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3b22119f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Variant AA Change</th>\n",
       "      <th>Source DB ID</th>\n",
       "      <th>Consequence Type</th>\n",
       "      <th>Clinical Significance</th>\n",
       "      <th>Phenotype/Disease</th>\n",
       "      <th>Simplified Significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NSRP1</td>\n",
       "      <td>p.Lys30Glu</td>\n",
       "      <td>RCV004545615</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>NSRP1-related disorder</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027</th>\n",
       "      <td>SERPINB6</td>\n",
       "      <td>p.Met5Ile</td>\n",
       "      <td>RCV001783729</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>Autosomal recessive nonsyndromic hearing loss 91</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1035</th>\n",
       "      <td>SERPINB6</td>\n",
       "      <td>p.Val45Ile</td>\n",
       "      <td>RCV004757137</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance, Likely benign</td>\n",
       "      <td>SERPINB6-related disorder</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1036</th>\n",
       "      <td>SERPINB6</td>\n",
       "      <td>p.Val45Ile</td>\n",
       "      <td>RCV001335154</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Variant of uncertain significance, Likely benign</td>\n",
       "      <td>Autosomal recessive nonsyndromic hearing loss 91</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>SERPINB6</td>\n",
       "      <td>p.Met94Val</td>\n",
       "      <td>RCV001807008</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Autosomal recessive nonsyndromic hearing loss 91</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687646</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asp132Val</td>\n",
       "      <td>RCV000004736</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic, Likely pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687716</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Val174Ile</td>\n",
       "      <td>RCV003133083</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely benign, Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687798</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asn224Ser</td>\n",
       "      <td>rs104894028</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687801</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asn224Ser</td>\n",
       "      <td>RCV000004743</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687866</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Gly275Arg</td>\n",
       "      <td>RCV000004745</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1347214 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene Name Variant AA Change  Source DB ID  Consequence Type  \\\n",
       "33           NSRP1        p.Lys30Glu  RCV004545615  missense variant   \n",
       "1027      SERPINB6         p.Met5Ile  RCV001783729  missense variant   \n",
       "1035      SERPINB6        p.Val45Ile  RCV004757137  missense variant   \n",
       "1036      SERPINB6        p.Val45Ile  RCV001335154  missense variant   \n",
       "1057      SERPINB6        p.Met94Val  RCV001807008  missense variant   \n",
       "...            ...               ...           ...               ...   \n",
       "35687646    NT5C3A       p.Asp132Val  RCV000004736  missense variant   \n",
       "35687716    NT5C3A       p.Val174Ile  RCV003133083  missense variant   \n",
       "35687798    NT5C3A       p.Asn224Ser   rs104894028  missense variant   \n",
       "35687801    NT5C3A       p.Asn224Ser  RCV000004743  missense variant   \n",
       "35687866    NT5C3A       p.Gly275Arg  RCV000004745  missense variant   \n",
       "\n",
       "                                     Clinical Significance  \\\n",
       "33                                           Likely benign   \n",
       "1027                                     Likely pathogenic   \n",
       "1035      Variant of uncertain significance, Likely benign   \n",
       "1036      Variant of uncertain significance, Likely benign   \n",
       "1057                                                Benign   \n",
       "...                                                    ...   \n",
       "35687646                     Pathogenic, Likely pathogenic   \n",
       "35687716  Likely benign, Variant of uncertain significance   \n",
       "35687798                                        Pathogenic   \n",
       "35687801                                        Pathogenic   \n",
       "35687866                                        Pathogenic   \n",
       "\n",
       "                                          Phenotype/Disease  \\\n",
       "33                                   NSRP1-related disorder   \n",
       "1027       Autosomal recessive nonsyndromic hearing loss 91   \n",
       "1035                              SERPINB6-related disorder   \n",
       "1036       Autosomal recessive nonsyndromic hearing loss 91   \n",
       "1057       Autosomal recessive nonsyndromic hearing loss 91   \n",
       "...                                                     ...   \n",
       "35687646  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687716  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687798                       Adenomas and Adenocarcinomas   \n",
       "35687801  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687866  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "\n",
       "         Simplified Significance  \n",
       "33                        Benign  \n",
       "1027                  Pathogenic  \n",
       "1035                      Benign  \n",
       "1036                      Benign  \n",
       "1057                      Benign  \n",
       "...                          ...  \n",
       "35687646              Pathogenic  \n",
       "35687716                  Benign  \n",
       "35687798              Pathogenic  \n",
       "35687801              Pathogenic  \n",
       "35687866              Pathogenic  \n",
       "\n",
       "[1347214 rows x 7 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Simplified Significance'].value_counts()\n",
    "\n",
    "df = df[\n",
    "    (df['Simplified Significance'].isin([\"Pathogenic\", \"Benign\"]))\n",
    "].copy()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9eec9562",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Variant AA Change</th>\n",
       "      <th>Source DB ID</th>\n",
       "      <th>Consequence Type</th>\n",
       "      <th>Clinical Significance</th>\n",
       "      <th>Phenotype/Disease</th>\n",
       "      <th>Simplified Significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31233377</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asp98Val</td>\n",
       "      <td>RCV000004736</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic, Likely pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31233446</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Val140Ile</td>\n",
       "      <td>RCV003133083</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely benign, Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31233518</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asn190Ser</td>\n",
       "      <td>RCV000004743</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31233574</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Gly241Arg</td>\n",
       "      <td>RCV000004745</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31233657</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asp86Val</td>\n",
       "      <td>RCV000004736</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic, Likely pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31233668</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Val128Ile</td>\n",
       "      <td>RCV003133083</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely benign, Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31233687</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asn178Ser</td>\n",
       "      <td>RCV000004743</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31233697</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Gly229Arg</td>\n",
       "      <td>RCV000004745</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687511</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Phe31Leu</td>\n",
       "      <td>RCV001802318</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Benign, Likely benign</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687646</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asp132Val</td>\n",
       "      <td>RCV000004736</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic, Likely pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687716</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Val174Ile</td>\n",
       "      <td>RCV003133083</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Likely benign, Variant of uncertain significance</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687798</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asn224Ser</td>\n",
       "      <td>rs104894028</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687801</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Asn224Ser</td>\n",
       "      <td>RCV000004743</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35687866</th>\n",
       "      <td>NT5C3A</td>\n",
       "      <td>p.Gly275Arg</td>\n",
       "      <td>RCV000004745</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Hemolytic anemia due to pyrimidine 5' nucleoti...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene Name Variant AA Change  Source DB ID  Consequence Type  \\\n",
       "31233377    NT5C3A        p.Asp98Val  RCV000004736  missense variant   \n",
       "31233446    NT5C3A       p.Val140Ile  RCV003133083  missense variant   \n",
       "31233518    NT5C3A       p.Asn190Ser  RCV000004743  missense variant   \n",
       "31233574    NT5C3A       p.Gly241Arg  RCV000004745  missense variant   \n",
       "31233657    NT5C3A        p.Asp86Val  RCV000004736  missense variant   \n",
       "31233668    NT5C3A       p.Val128Ile  RCV003133083  missense variant   \n",
       "31233687    NT5C3A       p.Asn178Ser  RCV000004743  missense variant   \n",
       "31233697    NT5C3A       p.Gly229Arg  RCV000004745  missense variant   \n",
       "35687511    NT5C3A        p.Phe31Leu  RCV001802318  missense variant   \n",
       "35687646    NT5C3A       p.Asp132Val  RCV000004736  missense variant   \n",
       "35687716    NT5C3A       p.Val174Ile  RCV003133083  missense variant   \n",
       "35687798    NT5C3A       p.Asn224Ser   rs104894028  missense variant   \n",
       "35687801    NT5C3A       p.Asn224Ser  RCV000004743  missense variant   \n",
       "35687866    NT5C3A       p.Gly275Arg  RCV000004745  missense variant   \n",
       "\n",
       "                                     Clinical Significance  \\\n",
       "31233377                     Pathogenic, Likely pathogenic   \n",
       "31233446  Likely benign, Variant of uncertain significance   \n",
       "31233518                                        Pathogenic   \n",
       "31233574                                        Pathogenic   \n",
       "31233657                     Pathogenic, Likely pathogenic   \n",
       "31233668  Likely benign, Variant of uncertain significance   \n",
       "31233687                                        Pathogenic   \n",
       "31233697                                        Pathogenic   \n",
       "35687511                             Benign, Likely benign   \n",
       "35687646                     Pathogenic, Likely pathogenic   \n",
       "35687716  Likely benign, Variant of uncertain significance   \n",
       "35687798                                        Pathogenic   \n",
       "35687801                                        Pathogenic   \n",
       "35687866                                        Pathogenic   \n",
       "\n",
       "                                          Phenotype/Disease  \\\n",
       "31233377  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "31233446  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "31233518  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "31233574  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "31233657  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "31233668  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "31233687  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "31233697  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687511  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687646  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687716  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687798                       Adenomas and Adenocarcinomas   \n",
       "35687801  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "35687866  Hemolytic anemia due to pyrimidine 5' nucleoti...   \n",
       "\n",
       "         Simplified Significance  \n",
       "31233377              Pathogenic  \n",
       "31233446                  Benign  \n",
       "31233518              Pathogenic  \n",
       "31233574              Pathogenic  \n",
       "31233657              Pathogenic  \n",
       "31233668                  Benign  \n",
       "31233687              Pathogenic  \n",
       "31233697              Pathogenic  \n",
       "35687511                  Benign  \n",
       "35687646              Pathogenic  \n",
       "35687716                  Benign  \n",
       "35687798              Pathogenic  \n",
       "35687801              Pathogenic  \n",
       "35687866              Pathogenic  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df[\"Gene Name\"] == \"NT5C3A\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7517f25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['missense variant', 'frameshift', 'nonsense', 'stop lost',\n",
       "       'stop gained', 'insertion', 'initiator codon variant',\n",
       "       'inframe deletion', 'delins', 'duplication', 'stop retained'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"Consequence Type\"].unique()\n",
    "['missense variant',\n",
    "       'stop gained',  \n",
    "       'inframe deletion', 'delins', 'duplication', 'stop retained']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0b7499fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo Symbol</th>\n",
       "      <th>Entrez Gene ID</th>\n",
       "      <th>GRCh37 Isoform</th>\n",
       "      <th>GRCh37 RefSeq</th>\n",
       "      <th>GRCh38 Isoform</th>\n",
       "      <th>GRCh38 RefSeq</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th># of occurrence within resources (Column J-P)</th>\n",
       "      <th>OncoKB Annotated</th>\n",
       "      <th>MSK-IMPACT</th>\n",
       "      <th>MSK-HEME</th>\n",
       "      <th>FOUNDATION ONE</th>\n",
       "      <th>FOUNDATION ONE HEME</th>\n",
       "      <th>Vogelstein</th>\n",
       "      <th>COSMIC CGC (v99)</th>\n",
       "      <th>Gene Aliases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>NM_005157.4</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>NM_005157.4</td>\n",
       "      <td>ONCOGENE</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ABL, JTK7, c-ABL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>207</td>\n",
       "      <td>ENST00000349310</td>\n",
       "      <td>NM_001014431.1</td>\n",
       "      <td>ENST00000349310</td>\n",
       "      <td>NM_001014431.1</td>\n",
       "      <td>ONCOGENE</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>AKT, PKB, PRKBA, RAC, RAC-alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALK</td>\n",
       "      <td>238</td>\n",
       "      <td>ENST00000389048</td>\n",
       "      <td>NM_004304.4</td>\n",
       "      <td>ENST00000389048</td>\n",
       "      <td>NM_004304.4</td>\n",
       "      <td>ONCOGENE</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CD246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMER1</td>\n",
       "      <td>139285</td>\n",
       "      <td>ENST00000330258</td>\n",
       "      <td>NM_152424.3</td>\n",
       "      <td>ENST00000374869</td>\n",
       "      <td>NM_152424.3</td>\n",
       "      <td>TSG</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>FAM123B, FLJ39827, RP11-403E24.2, WTX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC</td>\n",
       "      <td>324</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>TSG</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>DP2.5, PPP1R46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194</th>\n",
       "      <td>YY1</td>\n",
       "      <td>7528</td>\n",
       "      <td>ENST00000262238</td>\n",
       "      <td>NM_003403.4</td>\n",
       "      <td>ENST00000262238</td>\n",
       "      <td>NM_003403.4</td>\n",
       "      <td>ONCOGENE</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>INO80S, UCRBP, YIN-YANG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>ZFP36L1</td>\n",
       "      <td>677</td>\n",
       "      <td>ENST00000336440</td>\n",
       "      <td>NM_001244698.1</td>\n",
       "      <td>ENST00000336440</td>\n",
       "      <td>NM_001244698.1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Berg36, RNF162B, TIS11B, cMG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1198</th>\n",
       "      <td>ZFP36L2</td>\n",
       "      <td>678</td>\n",
       "      <td>ENST00000282388</td>\n",
       "      <td>NM_006887.4</td>\n",
       "      <td>ENST00000282388</td>\n",
       "      <td>NM_006887.4</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>ERF2, RNF162C, TIS11D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>ZNF292</td>\n",
       "      <td>23036</td>\n",
       "      <td>ENST00000369577</td>\n",
       "      <td>NM_015021.1</td>\n",
       "      <td>ENST00000369577</td>\n",
       "      <td>NM_015021.3</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>KIAA0530, ZFP292, Zn-15, Zn-16, bA393I2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>ZNF750</td>\n",
       "      <td>79755</td>\n",
       "      <td>ENST00000269394</td>\n",
       "      <td>NM_024702.2</td>\n",
       "      <td>ENST00000269394</td>\n",
       "      <td>NM_024702.2</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>FLJ13841, Zfp750</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>734 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo Symbol  Entrez Gene ID   GRCh37 Isoform   GRCh37 RefSeq  \\\n",
       "0           ABL1              25  ENST00000318560     NM_005157.4   \n",
       "1           AKT1             207  ENST00000349310  NM_001014431.1   \n",
       "2            ALK             238  ENST00000389048     NM_004304.4   \n",
       "3          AMER1          139285  ENST00000330258     NM_152424.3   \n",
       "4            APC             324  ENST00000257430     NM_000038.5   \n",
       "...          ...             ...              ...             ...   \n",
       "1194         YY1            7528  ENST00000262238     NM_003403.4   \n",
       "1197     ZFP36L1             677  ENST00000336440  NM_001244698.1   \n",
       "1198     ZFP36L2             678  ENST00000282388     NM_006887.4   \n",
       "1200      ZNF292           23036  ENST00000369577     NM_015021.1   \n",
       "1202      ZNF750           79755  ENST00000269394     NM_024702.2   \n",
       "\n",
       "       GRCh38 Isoform   GRCh38 RefSeq Gene Type  \\\n",
       "0     ENST00000318560     NM_005157.4  ONCOGENE   \n",
       "1     ENST00000349310  NM_001014431.1  ONCOGENE   \n",
       "2     ENST00000389048     NM_004304.4  ONCOGENE   \n",
       "3     ENST00000374869     NM_152424.3       TSG   \n",
       "4     ENST00000257430     NM_000038.5       TSG   \n",
       "...               ...             ...       ...   \n",
       "1194  ENST00000262238     NM_003403.4  ONCOGENE   \n",
       "1197  ENST00000336440  NM_001244698.1       TSG   \n",
       "1198  ENST00000282388     NM_006887.4       TSG   \n",
       "1200  ENST00000369577     NM_015021.3       TSG   \n",
       "1202  ENST00000269394     NM_024702.2       TSG   \n",
       "\n",
       "      # of occurrence within resources (Column J-P) OncoKB Annotated  \\\n",
       "0                                                 7              Yes   \n",
       "1                                                 7              Yes   \n",
       "2                                                 7              Yes   \n",
       "3                                                 7              Yes   \n",
       "4                                                 7              Yes   \n",
       "...                                             ...              ...   \n",
       "1194                                              1              Yes   \n",
       "1197                                              1              Yes   \n",
       "1198                                              1              Yes   \n",
       "1200                                              1              Yes   \n",
       "1202                                              1              Yes   \n",
       "\n",
       "     MSK-IMPACT MSK-HEME FOUNDATION ONE FOUNDATION ONE HEME Vogelstein  \\\n",
       "0           Yes      Yes            Yes                 Yes        Yes   \n",
       "1           Yes      Yes            Yes                 Yes        Yes   \n",
       "2           Yes      Yes            Yes                 Yes        Yes   \n",
       "3           Yes      Yes            Yes                 Yes        Yes   \n",
       "4           Yes      Yes            Yes                 Yes        Yes   \n",
       "...         ...      ...            ...                 ...        ...   \n",
       "1194         No       No             No                  No         No   \n",
       "1197         No       No             No                  No         No   \n",
       "1198         No       No             No                  No         No   \n",
       "1200         No       No             No                  No         No   \n",
       "1202         No       No             No                  No         No   \n",
       "\n",
       "     COSMIC CGC (v99)                               Gene Aliases  \n",
       "0                 Yes                           ABL, JTK7, c-ABL  \n",
       "1                 Yes            AKT, PKB, PRKBA, RAC, RAC-alpha  \n",
       "2                 Yes                                      CD246  \n",
       "3                 Yes      FAM123B, FLJ39827, RP11-403E24.2, WTX  \n",
       "4                 Yes                             DP2.5, PPP1R46  \n",
       "...               ...                                        ...  \n",
       "1194               No                  INO80S, UCRBP, YIN-YANG-1  \n",
       "1197               No              Berg36, RNF162B, TIS11B, cMG1  \n",
       "1198               No                      ERF2, RNF162C, TIS11D  \n",
       "1200               No  KIAA0530, ZFP292, Zn-15, Zn-16, bA393I2.3  \n",
       "1202               No                           FLJ13841, Zfp750  \n",
       "\n",
       "[734 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"/home/gdallagl/myworkdir/ESMSec/data/cell_cycle/datasets/3_parties/oncokb.org-cancer-genes_cancerGeneList.tsv\", sep=\"\\t\")\n",
    "df = df[df[\"Gene Type\"].isin([\"ONCOGENE\", \"TSG\"])]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7702837d",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7297358",
   "metadata": {},
   "source": [
    "# ------------------------------\n",
    "# ------------------------------\n",
    "# ------------------------------\n",
    "# ------------------------------\n",
    "# ------------------------------\n",
    "# ------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a6701a8",
   "metadata": {},
   "source": [
    "# YU dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f87c1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "# Parse all sequences in the FASTA file\n",
    "records = list(SeqIO.parse(\"/home/gdallagl/myworkdir/ESMSec/data/cell_cycle/0_class_yu_et_all.fasta\", \"fasta\"))\n",
    "data = {\n",
    "    \"protein\": [record.id for record in records],\n",
    "    \"sequence\": [str(record.seq) for record in records],\n",
    "    \"label\": [0 for record in records],\n",
    "}\n",
    "df_neg = pd.DataFrame(data)\n",
    "print(df_neg.shape)\n",
    "\n",
    "\n",
    "records = list(SeqIO.parse(\"/home/gdallagl/myworkdir/ESMSec/data/cell_cycle/1_class_yu_et_all.fasta\", \"fasta\"))\n",
    "data = {\n",
    "    \"protein\": [record.id for record in records],\n",
    "    \"sequence\": [str(record.seq) for record in records],\n",
    "    \"label\": [1 for record in records],\n",
    "}\n",
    "df_pos = pd.DataFrame(data)\n",
    "print(df_pos.shape)\n",
    "\n",
    "df = pd.concat([df_neg, df_pos])\n",
    "\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c97fbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_idx, test_idx = train_test_split(\n",
    "    df.index,\n",
    "    test_size=0.2,\n",
    "    stratify=df['label'],\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# Assign splits\n",
    "df['set'] = ''\n",
    "df.loc[train_idx, 'set'] = 'train'\n",
    "df.loc[test_idx, 'set'] = 'test'\n",
    "df.loc[test_idx[0], 'set'] = 'val' #just temp\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "df[\"protein\"] = range(df[\"protein\"].shape[0])\n",
    "\n",
    "print(df.groupby('set')['label'].value_counts().unstack(fill_value=0))\n",
    "display(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c6ad159",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.iloc[[332]].sequence.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df27ae82",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"/home/gdallagl/myworkdir/ESMSec/data/cell_cycle/yu_et_all_cyclins.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53b9bb34",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 2. Use pd.read_csv with appropriate parameters\n",
    "df = pd.read_csv(\n",
    "    \"/home/gdallagl/myworkdir/ESMSec/data/pfam/pdb_pfam_mapping.txt\", \n",
    "    sep='\\t', \n",
    "    comment='#', # Skips lines that start with '#' (the metadata lines)\n",
    "    skipinitialspace=True # Handles any extra spaces after the tab delimiter\n",
    ")\n",
    "\n",
    "# --- Explanation of the Output ---\n",
    "print(df.head())\n",
    "print(f\"\\nDataFrame shape: {df.shape}\")\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c545532",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- 2. Select the necessary columns ---\n",
    "# The PDB information is redundant if you only want protein-Pfam mapping\n",
    "df_mapping = df[['UNIPROT_ACCESSION', 'PFAM_ACCESSION']].copy()\n",
    "\n",
    "# --- 3. Filter for unique protein-domain pairs ---\n",
    "# Remove duplicates (e.g., if the same domain is found in multiple chains/PDB structures)\n",
    "df_unique = df_mapping.drop_duplicates()\n",
    "\n",
    "# --- 4. Group by Protein and Aggregate Pfam IDs ---\n",
    "# Group by the UniProt ID and aggregate all unique Pfam IDs into a list (or string)\n",
    "\n",
    "# Option A: Aggregate into a list (better for analysis)\n",
    "df_result_list = df_unique.groupby('UNIPROT_ACCESSION')['PFAM_ACCESSION'].agg(list).reset_index()\n",
    "df_result_list.columns = ['Protein_Accession', 'Pfam_Families_List']\n",
    "\n",
    "# Option B: Aggregate into a single string separated by a semicolon (easier for saving/display)\n",
    "df_result_string = df_unique.groupby('UNIPROT_ACCESSION')['PFAM_ACCESSION'].agg(lambda x: ';'.join(sorted(x))).reset_index()\n",
    "df_result_string.columns = ['Protein_Accession', 'Pfam_Families_String']\n",
    "\n",
    "\n",
    "# --- Output the preferred DataFrame ---\n",
    "print(\"--- DataFrame with Pfam Families as a List per Protein (Option A) ---\")\n",
    "print(df_result_list.head())\n",
    "print(f\"\\nTotal unique proteins: {len(df_result_list)}\")\n",
    "\n",
    "print(\"\\n--- DataFrame with Pfam Families as a Semicolon-Separated String (Option B) ---\")\n",
    "print(df_result_string.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eacdc32",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot = pd.read_csv(\"/home/gdallagl/myworkdir/ESMSec/data/UniRef50/human_proteome.tsv\", sep=\"\\t\")\n",
    "prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3bde9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_string = df_result_string[df_result_string.Protein_Accession.isin(set(prot[prot.Reviewed == \"reviewed\"].Entry))]\n",
    "\n",
    "df_result_string"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esm2_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
