{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "528107b0",
   "metadata": {},
   "source": [
    "# Make dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6e9348a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working directory successfully changed to: /home/gdallagl/myworkdir/ESMSec\n",
      "Seeds set successfully to 42 for random, numpy, and PyTorch (if used).\n"
     ]
    }
   ],
   "source": [
    "# Autorelaod\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "import igraph as ig\n",
    "import leidenalg\n",
    "from scipy.special import softmax\n",
    "import re\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import ast\n",
    "\n",
    "os.chdir(\"/home/gdallagl/myworkdir/ESMSec\")\n",
    "print(f\"Working directory successfully changed to: {os.getcwd()}\")\n",
    "\n",
    "import utils.my_functions as mf\n",
    "import utils.dataset_functions as dataf\n",
    "\n",
    "config_file=\"/home/gdallagl/myworkdir/ESMSec/config/config.yaml\"\n",
    "config = mf.load_config(config_file)\n",
    "\n",
    "mf.set_all_seeds(config[\"random_seed\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b08afced",
   "metadata": {},
   "source": [
    "# Load Uniprot proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2a8c41df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Gene Names (primary)</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Subcellular location [CC]</th>\n",
       "      <th>Gene Ontology (cellular component)</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>IsIsoform</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>TTN</td>\n",
       "      <td>TTN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>34350</td>\n",
       "      <td>SUBCELLULAR LOCATION: Cytoplasm {ECO:0000305|P...</td>\n",
       "      <td>condensed nuclear chromosome [GO:0000794]; cyt...</td>\n",
       "      <td>MTTQAPTFTQPLQSVVVLEGSTATFEAHISGFPVPEVSWFRDGQVI...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q8WXI7</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>MUC16</td>\n",
       "      <td>MUC16 CA125</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>14507</td>\n",
       "      <td>SUBCELLULAR LOCATION: Cell membrane; Single-pa...</td>\n",
       "      <td>external side of plasma membrane [GO:0009897];...</td>\n",
       "      <td>MLKPSGLPGSSSPTRSLMTGSRSTKATPEMDSGLTGATLSPKTSTG...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9H195</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>MUC3B</td>\n",
       "      <td>MUC3B</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>13477</td>\n",
       "      <td>SUBCELLULAR LOCATION: Membrane {ECO:0000305}; ...</td>\n",
       "      <td>Golgi lumen [GO:0005796]; plasma membrane [GO:...</td>\n",
       "      <td>MQLLGLLSILWMLKSSPGATGTLSTATSTSHVTFPRAEATRTALSN...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry  Reviewed Gene Names (primary)   Gene Names              Organism  \\\n",
       "0  Q8WZ42  reviewed                  TTN          TTN  Homo sapiens (Human)   \n",
       "1  Q8WXI7  reviewed                MUC16  MUC16 CA125  Homo sapiens (Human)   \n",
       "2  Q9H195  reviewed                MUC3B        MUC3B  Homo sapiens (Human)   \n",
       "\n",
       "   Length                          Subcellular location [CC]  \\\n",
       "0   34350  SUBCELLULAR LOCATION: Cytoplasm {ECO:0000305|P...   \n",
       "1   14507  SUBCELLULAR LOCATION: Cell membrane; Single-pa...   \n",
       "2   13477  SUBCELLULAR LOCATION: Membrane {ECO:0000305}; ...   \n",
       "\n",
       "                  Gene Ontology (cellular component)  \\\n",
       "0  condensed nuclear chromosome [GO:0000794]; cyt...   \n",
       "1  external side of plasma membrane [GO:0009897];...   \n",
       "2  Golgi lumen [GO:0005796]; plasma membrane [GO:...   \n",
       "\n",
       "                                            Sequence  IsIsoform  \n",
       "0  MTTQAPTFTQPLQSVVVLEGSTATFEAHISGFPVPEVSWFRDGQVI...      False  \n",
       "1  MLKPSGLPGSSSPTRSLMTGSRSTKATPEMDSGLTGATLSPKTSTG...      False  \n",
       "2  MQLLGLLSILWMLKSSPGATGTLSTATSTSHVTFPRAEATRTALSN...      False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number allowed proteins:  20210\n",
      "P0DMN0 O60343 Q86XX4 P48549 A0A1W2PPH5 P22460 Q8NHM4 A5PLN9 Q6UXL0 P23471 ...\n"
     ]
    }
   ],
   "source": [
    "proteome_uniprot = pd.read_csv(config[\"data_paths\"][\"human_proteome_uniprot\"])\n",
    "display(proteome_uniprot.head(3))\n",
    "\n",
    "# Define proteins to use in the follwing analasys\n",
    "ALLOWED_PROTEINS = set(proteome_uniprot.Entry)\n",
    "print(\"Number allowed proteins: \", len(ALLOWED_PROTEINS))\n",
    "print(*list(ALLOWED_PROTEINS)[:10], \"...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aba0adc2",
   "metadata": {},
   "source": [
    "# Load definitive positive genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d3fb6328",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nucleus-speckle_Uniprot-locations\n",
      "Positive genes df with mapped proteins\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>definitive_positive</th>\n",
       "      <th>geneset_count</th>\n",
       "      <th>Uniprot-locations</th>\n",
       "      <th>Gene Names (primary)</th>\n",
       "      <th>protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRRM2</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'nucleus speckle', 'nucleus'}</td>\n",
       "      <td>SRRM2</td>\n",
       "      <td>Q9UQ35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TENM1</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1</td>\n",
       "      <td>{'cytoplasm', 'nucleus speckle', 'cytoskeleton...</td>\n",
       "      <td>TENM1</td>\n",
       "      <td>Q9UKZ4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SON</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'nucleus speckle'}</td>\n",
       "      <td>SON</td>\n",
       "      <td>P18583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRPF8</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'nucleus speckle', 'nucleus'}</td>\n",
       "      <td>PRPF8</td>\n",
       "      <td>Q6P2Q9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASCC3</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'cytosol', 'nucleus speckle', 'cytoplasm', 'n...</td>\n",
       "      <td>ASCC3</td>\n",
       "      <td>Q8N3C0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>NXT1</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'nucleus speckle', 'cytoplasm', 'nucleus'}</td>\n",
       "      <td>NXT1</td>\n",
       "      <td>Q9UKK6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>CBY1</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'cytoplasm', 'cilium', 'flagellum', 'golgi ap...</td>\n",
       "      <td>CBY1</td>\n",
       "      <td>Q9Y3M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>PHF5A</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'nucleus speckle', 'nucleus'}</td>\n",
       "      <td>PHF5A</td>\n",
       "      <td>Q7RTV0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>TCIM</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'nucleolus', 'cytoplasm', 'nucleus speckle', ...</td>\n",
       "      <td>TCIM</td>\n",
       "      <td>Q9NR00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>SUMO1</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>{'cytoplasm', 'nucleus membrane', 'nucleus spe...</td>\n",
       "      <td>SUMO1</td>\n",
       "      <td>P63165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>203 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene definitive_positive  geneset_count  \\\n",
       "0    SRRM2            positive              1   \n",
       "1    TENM1           ambiguous              1   \n",
       "2      SON            positive              1   \n",
       "3    PRPF8            positive              1   \n",
       "4    ASCC3            positive              1   \n",
       "..     ...                 ...            ...   \n",
       "198   NXT1            positive              1   \n",
       "199   CBY1            positive              1   \n",
       "200  PHF5A            positive              1   \n",
       "201   TCIM            positive              1   \n",
       "202  SUMO1            positive              1   \n",
       "\n",
       "                                     Uniprot-locations Gene Names (primary)  \\\n",
       "0                       {'nucleus speckle', 'nucleus'}                SRRM2   \n",
       "1    {'cytoplasm', 'nucleus speckle', 'cytoskeleton...                TENM1   \n",
       "2                                  {'nucleus speckle'}                  SON   \n",
       "3                       {'nucleus speckle', 'nucleus'}                PRPF8   \n",
       "4    {'cytosol', 'nucleus speckle', 'cytoplasm', 'n...                ASCC3   \n",
       "..                                                 ...                  ...   \n",
       "198        {'nucleus speckle', 'cytoplasm', 'nucleus'}                 NXT1   \n",
       "199  {'cytoplasm', 'cilium', 'flagellum', 'golgi ap...                 CBY1   \n",
       "200                     {'nucleus speckle', 'nucleus'}                PHF5A   \n",
       "201  {'nucleolus', 'cytoplasm', 'nucleus speckle', ...                 TCIM   \n",
       "202  {'cytoplasm', 'nucleus membrane', 'nucleus spe...                SUMO1   \n",
       "\n",
       "    protein  \n",
       "0    Q9UQ35  \n",
       "1    Q9UKZ4  \n",
       "2    P18583  \n",
       "3    Q6P2Q9  \n",
       "4    Q8N3C0  \n",
       "..      ...  \n",
       "198  Q9UKK6  \n",
       "199  Q9Y3M2  \n",
       "200  Q7RTV0  \n",
       "201  Q9NR00  \n",
       "202  P63165  \n",
       "\n",
       "[203 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number initial positive genes 187\n",
      "Number positive proteins 187\n",
      "Number ambiguos proteins 16\n"
     ]
    }
   ],
   "source": [
    "#positive_genes_path = config[\"data_paths\"][\"final_positive_genes\"]\n",
    "positive_genes_path = '/home/gdallagl/myworkdir/ESMSec/data/cell_location/nucleus-speckle_Uniprot-locations_02_positive_proteins.csv'\n",
    "positive_genes_df = pd.read_csv(positive_genes_path)\n",
    "INITIAL_POSITIVE_GENES = set(positive_genes_df[positive_genes_df.definitive_positive == \"positive\"].gene)\n",
    "\n",
    "# Save dataset name for later\n",
    "dataset_name = \"_\".join(positive_genes_path.split(\"/\")[-1].split(\"_\")[:2])\n",
    "print(dataset_name)\n",
    "\n",
    "# Give protein names to genes usiong unirpto df\n",
    "# REMOVE not allowed proteins\n",
    "positive_genes_df = positive_genes_df.merge(proteome_uniprot[[\"Gene Names (primary)\", \"Entry\"]], how=\"inner\", left_on=\"gene\", right_on=\"Gene Names (primary)\")\n",
    "positive_genes_df.rename(columns={\"Entry\": \"protein\"}, inplace=True)\n",
    "\n",
    "print(\"Positive genes df with mapped proteins\")\n",
    "display(positive_genes_df)\n",
    "\n",
    "# List of geenes\n",
    "POSITIVE_PROTEINS = set(positive_genes_df[positive_genes_df.definitive_positive == \"positive\"].protein)\n",
    "AMBIGUOS_PROTEINS = set(positive_genes_df[positive_genes_df.definitive_positive == \"ambiguous\"].protein)\n",
    "print(\"Number initial positive genes\", len(INITIAL_POSITIVE_GENES))\n",
    "print(\"Number positive proteins\", len(POSITIVE_PROTEINS))\n",
    "print(\"Number ambiguos proteins\", len(AMBIGUOS_PROTEINS))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d7af74c",
   "metadata": {},
   "source": [
    "# Mapping positive-protein --> number genesets\n",
    "\n",
    "For sampling prob later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8499529e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Q9UQ35': 1, 'P18583': 1, 'Q6P2Q9': 1, 'Q8N3C0': 1, 'P49750': 1, 'Q14966': 1, 'Q9UPS6': 1, 'Q9H2Y7': 1, 'Q14997': 1, 'Q69YN4': 1, 'O15047': 1, 'Q92766': 1, 'Q5T200': 1, 'Q8NI27': 1, 'Q14004': 1, 'Q9NYV4': 1, 'Q9UKV3': 1, 'P51816': 1, 'O75533': 1, 'O15357': 1, 'Q03112': 1, 'Q86Z02': 1, 'Q6PD62': 1, 'Q9H7Z3': 1, 'Q8IZL2': 1, 'Q96JK9': 1, 'Q96SB8': 1, 'Q8WWM7': 1, 'Q9P2N5': 1, 'P42285': 1, 'Q7L014': 1, 'Q92585': 1, 'Q9NRA8': 1, 'Q96T37': 1, 'Q15020': 1, 'O75400': 1, 'Q9Y2W1': 1, 'P52824': 1, 'O94906': 1, 'Q9NYF8': 1, 'Q9HCG8': 1, 'Q8IYB3': 1, 'Q13435': 1, 'Q8NDT2': 1, 'Q9H6E5': 1, 'Q6NWY9': 1, 'Q99814': 1, 'Q86YF9': 1, 'Q9BZJ0': 1, 'P49756': 1, 'Q16665': 1, 'Q00839': 1, 'O75152': 1, 'Q8TF01': 1, 'Q99459': 1, 'Q8WTT2': 1, 'Q68CJ6': 1, 'Q15459': 1, 'Q86TB9': 1, 'Q13627': 1, 'Q9H1I8': 1, 'Q13427': 1, 'Q8NHY2': 1, 'Q96MU7': 1, 'Q9H307': 1, 'O00213': 1, 'P23246': 1, 'O15234': 1, 'Q02040': 1, 'P78362': 1, 'Q5TEA6': 1, 'O43395': 1, 'Q9Y2N7': 1, 'Q9P0U4': 1, 'O75925': 1, 'Q9BTA9': 1, 'Q9ULX6': 1, 'Q86YP4': 1, 'Q9Y6X2': 1, 'P55199': 1, 'O75928': 1, 'Q9UBU9': 1, 'P17844': 1, 'Q68DK7': 1, 'Q9UIS9': 1, 'Q8WXI9': 1, 'O43781': 1, 'O95391': 1, 'Q86U44': 1, 'O60508': 1, 'Q99755': 1, 'Q5FBB7': 1, 'O00257': 1, 'Q9H0G5': 1, 'Q16630': 1, 'Q14498': 1, 'Q9NW75': 1, 'Q9UK58': 1, 'Q9BZE0': 1, 'Q8WXF1': 1, 'O43172': 1, 'Q96S94': 1, 'Q8N9N5': 1, 'O43660': 1, 'Q2TAY7': 1, 'Q96GM8': 1, 'Q9C0B1': 1, 'P61011': 1, 'Q12874': 1, 'Q86WT6': 1, 'Q8WWY3': 1, 'Q8IYM9': 1, 'Q08170': 1, 'Q75N03': 1, 'Q15233': 1, 'Q495M9': 1, 'Q86X95': 1, 'P08621': 1, 'O95232': 1, 'Q13838': 1, 'P11801': 1, 'Q9BY77': 1, 'Q5JVS0': 1, 'P38919': 1, 'Q12837': 1, 'Q06587': 1, 'Q8N9N2': 1, 'Q15007': 1, 'Q6P6C2': 1, 'Q9Y383': 1, 'O75554': 1, 'O43791': 1, 'Q969K3': 1, 'Q96NB3': 1, 'Q9BWF3': 1, 'P18848': 1, 'Q96J01': 1, 'Q13247': 1, 'O15541': 1, 'Q99633': 1, 'Q86W42': 1, 'Q5HYJ3': 1, 'P48729': 1, 'Q96IZ7': 1, 'O75319': 1, 'P36873': 1, 'Q96SQ7': 1, 'Q96QD9': 1, 'P48431': 1, 'Q9HC98': 1, 'Q86U42': 1, 'Q15287': 1, 'P50222': 1, 'Q9H190': 1, 'P57052': 1, 'Q9NWB6': 1, 'Q96JC9': 1, 'O60828': 1, 'O75494': 1, 'Q96PM5': 1, 'Q96CJ1': 1, 'Q86V81': 1, 'Q07955': 1, 'Q9Y3Y2': 1, 'Q01081': 1, 'O75940': 1, 'Q66PJ3': 1, 'Q9BWG6': 1, 'Q01130': 1, 'Q8WU68': 1, 'P06730': 1, 'P82979': 1, 'Q6I9Y2': 1, 'Q05066': 1, 'Q9NX38': 1, 'O43447': 1, 'Q9Y5S9': 1, 'Q15038': 1, 'P84103': 1, 'Q13526': 1, 'O00422': 1, 'P61326': 1, 'Q9UKK6': 1, 'Q9Y3M2': 1, 'Q7RTV0': 1, 'Q9NR00': 1, 'P63165': 1}\n"
     ]
    }
   ],
   "source": [
    "tmp = positive_genes_df[positive_genes_df[\"definitive_positive\"] == \"positive\"]\n",
    "positive_protein_freq_map = dict(zip(tmp[\"protein\"], tmp[\"geneset_count\"]))\n",
    "print(positive_protein_freq_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27f25383",
   "metadata": {},
   "source": [
    "# Load Uniref50 Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "48a09ef4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19012, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniRef50_Q86YD7</td>\n",
       "      <td>[Q86YD7, Q658T7, A0A8V8TPE2, A8MXJ8, A6NKC0, A...</td>\n",
       "      <td>[FAM90A1, FAM90A2P, FAM90A3, FAM90A5, FAM90A7,...</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>[Q86YD7, Q658T7, A0A8V8TPE2, A8MXJ8, A6NKC0, A...</td>\n",
       "      <td>[FAM90A1, FAM90A2P, FAM90A3, FAM90A5, FAM90A7,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniRef50_Q6R6M4</td>\n",
       "      <td>[Q7RTZ2, Q6R6M4, A6NCW0, A6NCW7, A8MUK1, P0C7H...</td>\n",
       "      <td>[USP17L1, USP17L2, USP17L3, USP17L4, USP17L5, ...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>[Q7RTZ2, Q6R6M4, A6NCW0, A6NCW7, A8MUK1, P0C7H...</td>\n",
       "      <td>[USP17L1, USP17L2, USP17L3, USP17L4, USP17L5, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniRef50_Q13066</td>\n",
       "      <td>[A6NGK3, Q4V321, Q6NT46, Q13066, Q4V326, P0DTW...</td>\n",
       "      <td>[GAGE10, GAGE13, GAGE2A, GAGE2B, GAGE2E, GAGE1...</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>[A6NGK3, Q4V321, Q6NT46, Q13066, Q4V326, P0DTW...</td>\n",
       "      <td>[GAGE10, GAGE13, GAGE2A, GAGE2B, GAGE2E, GAGE1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniRef50_Cluster                                            protein  \\\n",
       "0  UniRef50_Q86YD7  [Q86YD7, Q658T7, A0A8V8TPE2, A8MXJ8, A6NKC0, A...   \n",
       "1  UniRef50_Q6R6M4  [Q7RTZ2, Q6R6M4, A6NCW0, A6NCW7, A8MUK1, P0C7H...   \n",
       "2  UniRef50_Q13066  [A6NGK3, Q4V321, Q6NT46, Q13066, Q4V326, P0DTW...   \n",
       "\n",
       "                                                gene  n_proteins  n_genes  \\\n",
       "0  [FAM90A1, FAM90A2P, FAM90A3, FAM90A5, FAM90A7,...          22       22   \n",
       "1  [USP17L1, USP17L2, USP17L3, USP17L4, USP17L5, ...          19       19   \n",
       "2  [GAGE10, GAGE13, GAGE2A, GAGE2B, GAGE2E, GAGE1...          17       17   \n",
       "\n",
       "                                   proteins_positive  \\\n",
       "0  [Q86YD7, Q658T7, A0A8V8TPE2, A8MXJ8, A6NKC0, A...   \n",
       "1  [Q7RTZ2, Q6R6M4, A6NCW0, A6NCW7, A8MUK1, P0C7H...   \n",
       "2  [A6NGK3, Q4V321, Q6NT46, Q13066, Q4V326, P0DTW...   \n",
       "\n",
       "                                      genes_positive  \n",
       "0  [FAM90A1, FAM90A2P, FAM90A3, FAM90A5, FAM90A7,...  \n",
       "1  [USP17L1, USP17L2, USP17L3, USP17L4, USP17L5, ...  \n",
       "2  [GAGE10, GAGE13, GAGE2A, GAGE2B, GAGE2E, GAGE1...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load already preprocessed version\n",
    "uniref_df = pd.read_csv(config[\"data_paths\"][\"uniref_df_path\"])\n",
    "\n",
    "# Make columsn lsits and not jsut stirngs\n",
    "uniref_df[\"protein\"] = uniref_df[\"protein\"].apply(ast.literal_eval)\n",
    "uniref_df[\"gene\"] = uniref_df[\"gene\"].apply(ast.literal_eval)\n",
    "\n",
    "# Filter allowed poteins\n",
    "uniref_df[\"proteins_positive\"], uniref_df[\"genes_positive\"] = zip(*uniref_df.apply(\n",
    "    lambda row: (\n",
    "        [p for p, g in zip(row[\"protein\"], row[\"gene\"]) if p in ALLOWED_PROTEINS],\n",
    "        [g for p, g in zip(row[\"protein\"], row[\"gene\"]) if p in ALLOWED_PROTEINS],\n",
    "    ),\n",
    "    axis=1\n",
    "))\n",
    "\n",
    "print(uniref_df.shape)\n",
    "uniref_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb94d680",
   "metadata": {},
   "source": [
    "## Mark positive genes and clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cf8f31e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster_label\n",
      "negative    18826\n",
      "positive      186\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Add the 'label' column: 'positive' if at least one protein maps to a positive gene, otherwise 'negative'\n",
    "uniref_df[\"cluster_label\"] = uniref_df[\"protein\"].apply(\n",
    "    lambda protein_list:\n",
    "    \"positive\" if any(p in POSITIVE_PROTEINS for p in protein_list)\n",
    "    else \"negative\"\n",
    ")\n",
    "\n",
    "# plot how many postive genesets\n",
    "print(uniref_df[\"cluster_label\"].value_counts())#.sort_index().plot(kind=\"bar\", logy=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09227cb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>n_proteins_positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>UniRef50_Q15233</td>\n",
       "      <td>[Q15233, Q8WXF1]</td>\n",
       "      <td>[NONO, PSPC1]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[Q15233, Q8WXF1]</td>\n",
       "      <td>[NONO, PSPC1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13463</th>\n",
       "      <td>UniRef50_O60828</td>\n",
       "      <td>[O60828]</td>\n",
       "      <td>[PQBP1]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[O60828]</td>\n",
       "      <td>[PQBP1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16395</th>\n",
       "      <td>UniRef50_Q5HYJ3</td>\n",
       "      <td>[Q5HYJ3]</td>\n",
       "      <td>[FAM76B]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q5HYJ3]</td>\n",
       "      <td>[FAM76B]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15000</th>\n",
       "      <td>UniRef50_Q14966</td>\n",
       "      <td>[Q14966]</td>\n",
       "      <td>[ZNF638]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q14966]</td>\n",
       "      <td>[ZNF638]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14416</th>\n",
       "      <td>UniRef50_Q2TAY7</td>\n",
       "      <td>[Q2TAY7]</td>\n",
       "      <td>[SMU1]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q2TAY7]</td>\n",
       "      <td>[SMU1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster           protein           gene  n_proteins  n_genes  \\\n",
       "404    UniRef50_Q15233  [Q15233, Q8WXF1]  [NONO, PSPC1]           2        2   \n",
       "13463  UniRef50_O60828          [O60828]        [PQBP1]           1        1   \n",
       "16395  UniRef50_Q5HYJ3          [Q5HYJ3]       [FAM76B]           1        1   \n",
       "15000  UniRef50_Q14966          [Q14966]       [ZNF638]           1        1   \n",
       "14416  UniRef50_Q2TAY7          [Q2TAY7]         [SMU1]           1        1   \n",
       "\n",
       "      proteins_positive genes_positive cluster_label  n_genes_positive  \\\n",
       "404    [Q15233, Q8WXF1]  [NONO, PSPC1]      positive                 2   \n",
       "13463          [O60828]        [PQBP1]      positive                 1   \n",
       "16395          [Q5HYJ3]       [FAM76B]      positive                 1   \n",
       "15000          [Q14966]       [ZNF638]      positive                 1   \n",
       "14416          [Q2TAY7]         [SMU1]      positive                 1   \n",
       "\n",
       "       n_proteins_positive  \n",
       "404                      2  \n",
       "13463                    1  \n",
       "16395                    1  \n",
       "15000                    1  \n",
       "14416                    1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(19012, 10)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='n_genes_positive'>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAicAAAGrCAYAAAAbyjfjAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAAIi1JREFUeJzt3X9QVXX+x/HXFRNUfhhRIAoy64aKFiSCmm6CUcjuUNpUtjmF7GZjQWpkpbO76Ndth2ryxzRdl6kdo3Z0ZZuMttzsB6uL61L8cPDHmj8odEkDdUsQKrTL+f7RdIsVTYTlfg7n+Zi5M3vvOZ7zxmXX59zzy2VZliUAAABD9PP1AAAAAN9HnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKP19PUBXtbe369ixYwoKCpLL5fL1OAAA4CJYlqXTp08rMjJS/fpd+LsR28XJsWPHFBUV5esxAADAJaivr9fw4cMvuI5t4sTtdsvtduvrr7+W9M0PFxwc7OOpAADAxWhublZUVJSCgoJ+cF2X3W5f39zcrJCQEDU1NREnAADYRFf+/eaEWAAAYBTiBAAAGIU4AQAARiFOAACAUYgTAABgFOIEAAAYhTgBAABGIU4AAIBRiBMAAGAU4gQAABiFOAEAAEYhTgAAgFFsEydut1txcXFKSkry9SgAAOB/iKcSAwCA/7mu/Pvdv5dmcpyYJZt9PUKfcfjJn/l6BABAL7LNYR0AAOAMxAkAADAKcQIAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwCnECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIziszj54osvNGLECC1evNhXIwAAAAP5LE5+97vfadKkSb7aPQAAMJRP4uTQoUPav3+/MjIyfLF7AABgsC7HSVlZmTIzMxUZGSmXy6WSkpJz1nG73YqJiVFAQIAmTpyoioqKDssXL16sgoKCSx4aAAD0XV2Ok9bWVsXHx8vtdne6vLi4WHl5eVq2bJl27typ+Ph4paen6/jx45Kk119/XbGxsYqNje3e5AAAoE/q39U/kJGRccHDMatWrdK8efOUnZ0tSSosLNTmzZu1bt06LVmyRO+//742btyoV155RS0tLTp79qyCg4OVn5/f6fba2trU1tbmfd/c3NzVkQEAgI306DknZ86cUXV1tdLS0r7bQb9+SktLU3l5uSSpoKBA9fX1Onz4sJ555hnNmzfvvGHy7fohISHeV1RUVE+ODAAADNOjcXLy5El5PB6Fh4d3+Dw8PFwNDQ2XtM2lS5eqqanJ+6qvr++JUQEAgKG6fFinJ82dO/cH1/H395e/v///fhgAAGCEHv3mJCwsTH5+fmpsbOzweWNjoyIiIrq1bbfbrbi4OCUlJXVrOwAAwGw9GicDBgxQYmKiSktLvZ+1t7ertLRUkydP7ta2c3JytG/fPlVWVnZ3TAAAYLAuH9ZpaWlRbW2t931dXZ1qamoUGhqq6Oho5eXlKSsrSxMmTFBycrLWrFmj1tZW79U7AAAAF9LlOKmqqlJqaqr3fV5eniQpKytLRUVFmj17tk6cOKH8/Hw1NDQoISFBW7ZsOeckWQAAgM64LMuyfD3ExXC73XK73fJ4PDp48KCampoUHBzs67HOK2bJZl+P0GccfvJnvh4BANBNzc3NCgkJuah/v3324L+u4pwTAACcwTZxAgAAnIE4AQAARiFOAACAUWwTJ9yEDQAAZ7BNnHBCLAAAzmCbOAEAAM5AnAAAAKMQJwAAwCi2iRNOiAUAwBlsEyecEAsAgDPYJk4AAIAzECcAAMAoxAkAADAKcQIAAIximzjhah0AAJzBNnHC1ToAADiDbeIEAAA4A3ECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIximzjhPicAADiDbeKE+5wAAOAMtokTAADgDMQJAAAwCnECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIxCnAAAAKPYJk64QywAAM5gmzjhDrEAADiDbeIEAAA4A3ECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAo9gmTnjwHwAAzmCbOOHBfwAAOINt4gQAADgDcQIAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwCnECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIxCnAAAAKMQJwAAwCjECQAAMEqvx8mpU6c0YcIEJSQkaNy4cXrhhRd6ewQAAGCw/r29w6CgIJWVlWnQoEFqbW3VuHHjdNttt+mKK67o7VEAAICBev2bEz8/Pw0aNEiS1NbWJsuyZFlWb48BAAAM1eU4KSsrU2ZmpiIjI+VyuVRSUnLOOm63WzExMQoICNDEiRNVUVHRYfmpU6cUHx+v4cOH69FHH1VYWNgl/wAAAKBv6XKctLa2Kj4+Xm63u9PlxcXFysvL07Jly7Rz507Fx8crPT1dx48f964zZMgQ7dq1S3V1ddqwYYMaGxvPu7+2tjY1Nzd3eAEAgL6ry3GSkZGhJ554QrNmzep0+apVqzRv3jxlZ2crLi5OhYWFGjRokNatW3fOuuHh4YqPj9f27dvPu7+CggKFhIR4X1FRUV0dGQAA2EiPnnNy5swZVVdXKy0t7bsd9OuntLQ0lZeXS5IaGxt1+vRpSVJTU5PKyso0atSo825z6dKlampq8r7q6+t7cmQAAGCYHr1a5+TJk/J4PAoPD+/weXh4uPbv3y9JOnLkiO6//37vibAPPfSQrrnmmvNu09/fX/7+/j05JgAAMFivX0qcnJysmpqa3t4tAACwiR49rBMWFiY/P79zTnBtbGxUREREt7btdrsVFxenpKSkbm0HAACYrUfjZMCAAUpMTFRpaan3s/b2dpWWlmry5Mnd2nZOTo727dunysrK7o4JAAAM1uXDOi0tLaqtrfW+r6urU01NjUJDQxUdHa28vDxlZWVpwoQJSk5O1po1a9Ta2qrs7OweHRwAAPRNXY6Tqqoqpaamet/n5eVJkrKyslRUVKTZs2frxIkTys/PV0NDgxISErRly5ZzTpIFAADojMuyyb3j3W633G63PB6PDh48qKamJgUHB/t6rPOKWbLZ1yP0GYef/JmvRwAAdFNzc7NCQkIu6t/vXn+2zqXinBMAAJzBNnECAACcgTgBAABGsU2ccJ8TAACcwTZxwjknAAA4g23iBAAAOANxAgAAjEKcAAAAoxAnAADAKLaJE67WAQDAGWwTJ1ytAwCAM9gmTgAAgDMQJwAAwCjECQAAMApxAgAAjGKbOOFqHQAAnME2ccLVOgAAOINt4gQAADgDcQIAAIxCnAAAAKMQJwAAwCjECQAAMIpt4oRLiQEAcAbbxAmXEgMA4Ay2iRMAAOAMxAkAADAKcQIAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAo9gmTrhDLAAAzmCbOOEOsQAAOINt4gQAADgDcQIAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwCnECAACMQpwAAACjECcAAMAotokTnkoMAIAz2CZOeCoxAADOYJs4AQAAztDf1wMA6D0xSzb7eoQ+4fCTP/P1CECfxjcnAADAKMQJAAAwCnECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwSq/HSX19vVJSUhQXF6drr71Wr7zySm+PAAAADNa/13fYv7/WrFmjhIQENTQ0KDExUT/96U81ePDg3h4FAAAYqNfjZOjQoRo6dKgkKSIiQmFhYfrss8+IEwAAIOkSDuuUlZUpMzNTkZGRcrlcKikpOWcdt9utmJgYBQQEaOLEiaqoqOh0W9XV1fJ4PIqKiury4AAAoG/qcpy0trYqPj5ebre70+XFxcXKy8vTsmXLtHPnTsXHxys9PV3Hjx/vsN5nn32me++9V88///ylTQ4AAPqkLh/WycjIUEZGxnmXr1q1SvPmzVN2drYkqbCwUJs3b9a6deu0ZMkSSVJbW5tmzpypJUuW6Prrr7/g/tra2tTW1uZ939zc3NWRAQCAjfTo1TpnzpxRdXW10tLSvttBv35KS0tTeXm5JMmyLM2dO1fTp0/XPffc84PbLCgoUEhIiPfFISAAAPq2Ho2TkydPyuPxKDw8vMPn4eHhamhokCTt2LFDxcXFKikpUUJCghISErRnz57zbnPp0qVqamryvurr63tyZAAAYJhev1pn6tSpam9vv+j1/f395e/v/z+cCAAAmKRHvzkJCwuTn5+fGhsbO3ze2NioiIiIbm3b7XYrLi5OSUlJ3doOAAAwW4/GyYABA5SYmKjS0lLvZ+3t7SotLdXkyZO7te2cnBzt27dPlZWV3R0TAAAYrMuHdVpaWlRbW+t9X1dXp5qaGoWGhio6Olp5eXnKysrShAkTlJycrDVr1qi1tdV79Q4AAMCFdDlOqqqqlJqa6n2fl5cnScrKylJRUZFmz56tEydOKD8/Xw0NDUpISNCWLVvOOUkWAACgM12Ok5SUFFmWdcF1cnNzlZube8lDdcbtdsvtdsvj8fTodgEAgFl6/anEl4pzTgAAcAbbxAkAAHAG4gQAABiFOAEAAEaxTZxwEzYAAJzBNnHCCbEAADiDbeIEAAA4A3ECAACMQpwAAACj2CZOOCEWAABnsE2ccEIsAADOYJs4AQAAzkCcAAAAoxAnAADAKMQJAAAwim3ihKt1AABwBtvECVfrAADgDLaJEwAA4AzECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwim3ihPucAADgDLaJE+5zAgCAM9gmTgAAgDMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwCnECAACMYps44Q6xAAA4g23ihDvEAgDgDLaJEwAA4AzECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwCnECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIxCnAAAAKPYJk54KjEAAM5gmzjhqcQAADiDbeIEAAA4A3ECAACMQpwAAACjECcAAMAoxAkAADAKcQIAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKMQJAAAwCnECAACMQpwAAACjECcAAMAoPomTWbNm6fLLL9ftt9/ui90DAACD+SROFi5cqJdfftkXuwYAAIbzSZykpKQoKCjIF7sGAACG63KclJWVKTMzU5GRkXK5XCopKTlnHbfbrZiYGAUEBGjixImqqKjoiVkBAIADdDlOWltbFR8fL7fb3eny4uJi5eXladmyZdq5c6fi4+OVnp6u48ePX9KAbW1tam5u7vACAAB9V5fjJCMjQ0888YRmzZrV6fJVq1Zp3rx5ys7OVlxcnAoLCzVo0CCtW7fukgYsKChQSEiI9xUVFXVJ2wEAAPbQo+ecnDlzRtXV1UpLS/tuB/36KS0tTeXl5Ze0zaVLl6qpqcn7qq+v76lxAQCAgfr35MZOnjwpj8ej8PDwDp+Hh4dr//793vdpaWnatWuXWltbNXz4cL3yyiuaPHlyp9v09/eXv79/T44JAAAM1qNxcrHee+89X+wWAADYQI8e1gkLC5Ofn58aGxs7fN7Y2KiIiIhubdvtdisuLk5JSUnd2g4AADBbj8bJgAEDlJiYqNLSUu9n7e3tKi0tPe9hm4uVk5Ojffv2qbKysrtjAgAAg3X5sE5LS4tqa2u97+vq6lRTU6PQ0FBFR0crLy9PWVlZmjBhgpKTk7VmzRq1trYqOzu7RwcHAAB9U5fjpKqqSqmpqd73eXl5kqSsrCwVFRVp9uzZOnHihPLz89XQ0KCEhARt2bLlnJNkAQAAOtPlOElJSZFlWRdcJzc3V7m5uZc8VGfcbrfcbrc8Hk+PbhcAAJjFJ8/WuRSccwIAgDPYJk4AAIAzECcAAMAotokT7nMCAIAz2CZOOOcEAABnsE2cAAAAZyBOAACAUYgTAABgFOIEAAAYxTZxwtU6AAA4g23ihKt1AABwBtvECQAAcAbiBAAAGIU4AQAARiFOAACAUWwTJ1ytAwCAM9gmTrhaBwAAZ7BNnAAAAGcgTgAAgFGIEwAAYBTiBAAAGIU4AQAARiFOAACAUWwTJ9znBAAAZ7BNnHCfEwAAnME2cQIAAJyBOAEAAEYhTgAAgFGIEwAAYBTiBAAAGIU4AQAARiFOAACAUWwTJ9yEDQAAZ7BNnHATNgAAnME2cQIAAJyBOAEAAEYhTgAAgFGIEwAAYBTiBAAAGIU4AQAARiFOAACAUYgTAABgFOIEAAAYhTgBAABGIU4AAIBRiBMAAGAU28QJTyUGAMAZbBMnPJUYAABnsE2cAAAAZyBOAACAUYgTAABgFOIEAAAYhTgBAABGIU4AAIBRiBMAAGAU4gQAABiFOAEAAEYhTgAAgFGIEwAAYBTiBAAAGIU4AQAARiFOAACAUYgTAABgFOIEAAAYhTgBAABG8UmcvPnmmxo1apSuvvpq/eEPf/DFCAAAwFD9e3uHX3/9tfLy8rR161aFhIQoMTFRs2bN0hVXXNHbowAAAAP1+jcnFRUVGjt2rIYNG6bAwEBlZGTonXfe6e0xAACAobocJ2VlZcrMzFRkZKRcLpdKSkrOWcftdismJkYBAQGaOHGiKioqvMuOHTumYcOGed8PGzZMR48evbTpAQBAn9PlOGltbVV8fLzcbneny4uLi5WXl6dly5Zp586dio+PV3p6uo4fP35JA7a1tam5ubnDCwAA9F1dPuckIyNDGRkZ512+atUqzZs3T9nZ2ZKkwsJCbd68WevWrdOSJUsUGRnZ4ZuSo0ePKjk5+bzbKygo0P/93/91dUwAgA3ELNns6xH6jMNP/szXI/SYHj3n5MyZM6qurlZaWtp3O+jXT2lpaSovL5ckJScna+/evTp69KhaWlr01ltvKT09/bzbXLp0qZqamryv+vr6nhwZAAAYpkev1jl58qQ8Ho/Cw8M7fB4eHq79+/d/s8P+/bVy5Uqlpqaqvb1djz322AWv1PH395e/v39PjgkAAAzW65cSS9Itt9yiW265xRe7BgAAhuvRwzphYWHy8/NTY2Njh88bGxsVERHRrW273W7FxcUpKSmpW9sBAABm69E4GTBggBITE1VaWur9rL29XaWlpZo8eXK3tp2Tk6N9+/apsrKyu2MCAACDdfmwTktLi2pra73v6+rqVFNTo9DQUEVHRysvL09ZWVmaMGGCkpOTtWbNGrW2tnqv3gEAALiQLsdJVVWVUlNTve/z8vIkSVlZWSoqKtLs2bN14sQJ5efnq6GhQQkJCdqyZcs5J8kCAAB0pstxkpKSIsuyLrhObm6ucnNzL3mozrjdbrndbnk8nh7dLgAAMItPnkp8KTjnBAAAZ7BNnAAAAGcgTgAAgFFsEyfc5wQAAGewTZxwzgkAAM5gmzgBAADOQJwAAACj+OTBf93x7T1WmpubfTzJhbW3feHrEfoM0/+7thN+L3sGv5M9h9/JnmP67+W38/3QvdIkyWVdzFoG+eSTTxQVFeXrMQAAwCWor6/X8OHDL7iO7eKkvb1dx44dU1BQkFwul6/HsbXm5mZFRUWpvr5ewcHBvh4H4HcSxuF3sudYlqXTp08rMjJS/fpd+KwS2x3W6dev3w8WF7omODiY/9HBKPxOwjT8TvaMkJCQi1qPE2IBAIBRiBMAAGAU4sTB/P39tWzZMvn7+/t6FEASv5MwD7+TvmG7E2IBAEDfxjcnAADAKMQJAAAwCnECAACMQpwAAACjECcAAMAotrtDLC7dyZMntW7dOpWXl6uhoUGSFBERoeuvv15z587VlVde6eMJAQDgmxPHqKysVGxsrJ599lmFhITohhtu0A033KCQkBA9++yzGj16tKqqqnw9JtBBfX29fvGLX/h6DDjIl19+qX/84x/at2/fOcu++uorvfzyyz6Yynm4z4lDTJo0SfHx8SosLDzngYmWZWn+/PnavXu3ysvLfTQhcK5du3Zp/Pjx8ng8vh4FDnDw4EHdfPPN+ve//y2Xy6WpU6dq48aNGjp0qCSpsbFRkZGR/D72Ag7rOMSuXbtUVFTU6ZOcXS6XHn74YV133XU+mAxO9pe//OWCyz/++ONemgSQHn/8cY0bN05VVVU6deqUFi1apClTpmjbtm2Kjo729XiOQpw4REREhCoqKjR69OhOl1dUVCg8PLyXp4LTzZw5Uy6XSxf6ArezoAb+F/75z3/qvffeU1hYmMLCwvTGG2/owQcf1E9+8hNt3bpVgwcP9vWIjkGcOMTixYt1//33q7q6WjfeeKM3RBobG1VaWqoXXnhBzzzzjI+nhNMMHTpUa9eu1a233trp8pqaGiUmJvbyVHCqL7/8Uv37f/fPosvl0u9//3vl5uZq2rRp2rBhgw+ncxbixCFycnIUFham1atXa+3atd5jpn5+fkpMTFRRUZHuvPNOH08Jp0lMTFR1dfV54+SHvlUBetK3FwaMGTOmw+fPPfecJOmWW27xxViOxAmxDnT27FmdPHlSkhQWFqbLLrvMxxPBqbZv367W1lbNmDGj0+Wtra2qqqrStGnTenkyOFFBQYG2b9+uv/71r50uf/DBB1VYWKj29vZensx5iBMAAGAU7nMCAACMQpwAAACjECcAAMAoxAkAADAKcQIAklJSUrRo0aILrlNUVKQhQ4b0yjyAk3G1DgBI+uyzz3TZZZcpKChIkhQTE6NFixZ1CJYvv/xSp0+f1lVXXeWjKQFn4CZsACApNDT0B9cZOHCgBg4c2AvTAM7GYR3AJlJSUrRgwQI99thjCg0NVUREhJYvX35Rf3b//v2aOnWqAgICFBcXp/fee08ul0slJSXederr63XnnXdqyJAhCg0N1a233qrDhw97l8+dO1czZ87UM888o6FDh+qKK65QTk6Ozp49612nra1Nixcv1rBhwzR48GBNnDhR27Zt8y4/cuSIMjMzdfnll2vw4MEaO3bseW949X3btm2Ty+XS5s2bde211yogIECTJk3S3r17O6z36quvauzYsfL391dMTIxWrlzZYfnatWt19dVXKyAgQOHh4br99ts7/P1++y1JSkqKjhw5oocfflgul8v7fJ/vH9Y5ePCgXC6X9u/f32Efq1ev1siRI73v9+7dq4yMDAUGBio8PFz33HOP9yaIADpHnAA28tJLL2nw4MH64IMP9PTTT2vFihV69913L/hnPB6PZs6cqUGDBumDDz7Q888/r1/96lcd1jl79qzS09MVFBSk7du3a8eOHQoMDNSMGTN05swZ73pbt27VRx99pK1bt+qll15SUVGRioqKvMtzc3NVXl6ujRs3avfu3brjjjs0Y8YMHTp0SNI3j1Foa2tTWVmZ9uzZo6eeekqBgYEX/fM/+uijWrlypSorK3XllVcqMzPTG0fV1dW68847ddddd2nPnj1avny5fvOb33jnq6qq0oIFC7RixQodOHBAW7Zs0Q033NDpfjZt2qThw4drxYoV+vTTT/Xpp5+es05sbKwmTJig9evXd/h8/fr1uvvuuyVJp06d0vTp03XdddepqqpKW7ZsUWNjI4+KAH6IBcAWpk2bZk2dOrXDZ0lJSdbjjz9+wT/31ltvWf3797c+/fRT72fvvvuuJcl67bXXLMuyrD/+8Y/WqFGjrPb2du86bW1t1sCBA623337bsizLysrKskaMGGF9/fXX3nXuuOMOa/bs2ZZlWdaRI0csPz8/6+jRox32f+ONN1pLly61LMuyrrnmGmv58uVd/Mkta+vWrZYka+PGjd7P/vOf/1gDBw60iouLLcuyrLvvvtu66aabOvy5Rx991IqLi7Msy7JeffVVKzg42Gpubu50H9OmTbMWLlzofT9ixAhr9erVHdZ58cUXrZCQEO/71atXWyNHjvS+P3DggCXJ+vDDDy3Lsqzf/va31s0339xhG/X19ZYk68CBAxf3wwMOxDcngI1ce+21Hd4PHTpUx48fv+CfOXDggKKiohQREeH9LDk5ucM6u3btUm1trYKCghQYGKjAwECFhobqq6++0kcffeRdb+zYsfLz8+t0/3v27JHH41FsbKx3G4GBgfr73//u3caCBQv0xBNPaMqUKVq2bJl2797dpZ9/8uTJ3v8cGhqqUaNG6cMPP5Qkffjhh5oyZUqH9adMmaJDhw7J4/Hopptu0ogRI/SjH/1I99xzj9avX68vvviiS/v/b3fddZcOHz6s999/X9I335qMHz9eo0ePlvTN3+vWrVs7/H18u+z7f68AOuKEWMBG/vshjS6Xq0ceQtbS0qLExMRzDlFI0pVXXnlR+29paZGfn5+qq6s7BIwk76Gb++67T+np6dq8ebPeeecdFRQUaOXKlXrooYe6/TP8kKCgIO3cuVPbtm3TO++8o/z8fC1fvlyVlZWXfHlwRESEpk+frg0bNmjSpEnasGGDHnjgAe/ylpYWZWZm6qmnnjrnzw4dOvRSfxSgz+ObE6CPGzVqlOrr69XY2Oj9rLKyssM648eP16FDh3TVVVfpxz/+cYdXSEjIRe3nuuuuk8fj0fHjx8/Zxve/tYmKitL8+fO1adMmPfLII3rhhRcu+mf59hsKSfr888918OBB7+Ptx4wZox07dnRYf8eOHYqNjfXGUv/+/ZWWlqann35au3fv1uHDh/W3v/2t030NGDBAHo/nB2eaM2eOiouLVV5ero8//lh33XWXd9n48eP1r3/9SzExMef8nQwePPiif27AaYgToI+76aabNHLkSGVlZWn37t3asWOHfv3rX0uS9yqUOXPmKCwsTLfeequ2b9+uuro6bdu2TQsWLNAnn3xyUfuJjY3VnDlzdO+992rTpk2qq6tTRUWFCgoKtHnzZknSokWL9Pbbb6uurk47d+7U1q1bvXFxMVasWKHS0lLt3btXc+fOVVhYmGbOnClJeuSRR1RaWqrf/va3OnjwoF566SU999xzWrx4sSTpzTff1LPPPquamhodOXJEL7/8strb2zVq1KhO9xUTE6OysjIdPXr0glfX3HbbbTp9+rQeeOABpaamKjIy0rssJydHn332mX7+85+rsrJSH330kd5++21lZ2dfVPgATkWcAH2cn5+fSkpK1NLSoqSkJN13333eq3UCAgIkSYMGDVJZWZmio6N12223acyYMfrlL3+pr776SsHBwRe9rxdffFH33nuvHnnkEY0aNUozZ85UZWWloqOjJX1z5VBOTo7GjBmjGTNmKDY2VmvXrr3o7T/55JNauHChEhMT1dDQoDfeeEMDBgyQ9M23FH/+85+1ceNGjRs3Tvn5+VqxYoXmzp0rSRoyZIg2bdqk6dOna8yYMSosLNSf/vQnjR07ttN9rVixQocPH9bIkSM7HNr6b0FBQcrMzNSuXbs0Z86cDssiIyO1Y8cOeTwe3Xzzzbrmmmu0aNEiDRkyRP368X+/wPlwh1jAgXbs2KGpU6eqtra2wz05TLVt2zalpqbq888/5/bxgANwQizgAK+99poCAwN19dVXq7a2VgsXLtSUKVNsESYAnIfvFQGbW79+fYdLVb//+vaQxenTp5WTk6PRo0dr7ty5SkpK0uuvv+7jyb8zf/788/4M8+fP9/V4AHoZh3UAmzt9+nSHK3G+77LLLtOIESN6eaKuO378uJqbmztdFhwczIP2AIchTgAAgFE4rAMAAIxCnAAAAKMQJwAAwCjECQAAMApxAgAAjEKcAAAAoxAnAADAKP8PHGF949lb9HkAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create list of genes AND CORREPSONDET prots that are postive (from above)\n",
    "# NEED POSITIVE GENES FOR ENRICHEMENT ANALSYS\n",
    "\n",
    "# unzip in parallel pritens_clened and genes_clened and create ritens_clened_psotive and genes_clened_positive \n",
    "    # maintain if proteins comes form psotive gene\n",
    "uniref_df[\"proteins_positive\"], uniref_df[\"genes_positive\"] = zip(*uniref_df.apply(\n",
    "    lambda row: (\n",
    "        [p for p, g in zip(row[\"protein\"], row[\"gene\"]) if p in POSITIVE_PROTEINS], # protein is postive based on gene\n",
    "        [g for p, g in zip(row[\"protein\"], row[\"gene\"]) if p in POSITIVE_PROTEINS], # protein is postive based on gene\n",
    "    ),\n",
    "    axis=1\n",
    "))\n",
    "uniref_df[\"n_genes_positive\"] = uniref_df[\"genes_positive\"].apply(len)\n",
    "uniref_df[\"n_proteins_positive\"] = uniref_df[\"proteins_positive\"].apply(len)\n",
    "uniref_df.sort_values(by=\"n_genes_positive\", ascending=False, inplace=True)\n",
    "\n",
    "display(uniref_df.head(5))\n",
    "display(uniref_df.shape)\n",
    "uniref_df.n_genes_positive.value_counts().plot(kind=\"bar\", logy=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7631505",
   "metadata": {},
   "source": [
    "## Give probs to positive genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f3bd6da3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>n_proteins_positive</th>\n",
       "      <th>logits</th>\n",
       "      <th>probs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>UniRef50_Q15233</td>\n",
       "      <td>[Q15233, Q8WXF1]</td>\n",
       "      <td>[NONO, PSPC1]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[Q15233, Q8WXF1]</td>\n",
       "      <td>[NONO, PSPC1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[1, 1]</td>\n",
       "      <td>[0.5, 0.5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13463</th>\n",
       "      <td>UniRef50_O60828</td>\n",
       "      <td>[O60828]</td>\n",
       "      <td>[PQBP1]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[O60828]</td>\n",
       "      <td>[PQBP1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster           protein           gene  n_proteins  n_genes  \\\n",
       "404    UniRef50_Q15233  [Q15233, Q8WXF1]  [NONO, PSPC1]           2        2   \n",
       "13463  UniRef50_O60828          [O60828]        [PQBP1]           1        1   \n",
       "\n",
       "      proteins_positive genes_positive cluster_label  n_genes_positive  \\\n",
       "404    [Q15233, Q8WXF1]  [NONO, PSPC1]      positive                 2   \n",
       "13463          [O60828]        [PQBP1]      positive                 1   \n",
       "\n",
       "       n_proteins_positive  logits       probs  \n",
       "404                      2  [1, 1]  [0.5, 0.5]  \n",
       "13463                    1     [1]       [1.0]  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# give probablities to be samped to each protein\n",
    "uniref_df[\"logits\"] = uniref_df[\"proteins_positive\"].apply(\n",
    "    lambda gene_list: [positive_protein_freq_map.get(g, 0) for g in gene_list]\n",
    ")\n",
    "def safe_softmax(logits):\n",
    "    if len(logits) == 0:\n",
    "        return []  # return empty list if no logits\n",
    "    return softmax(logits).tolist()  # convert numpy array to list\n",
    "uniref_df[\"probs\"] = uniref_df[\"logits\"].apply(safe_softmax)\n",
    "\n",
    "uniref_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1198c8e6",
   "metadata": {},
   "source": [
    "# Mark NEGATIVE proteins\n",
    "\n",
    "Create a list for each cluster that contains genes that are negative. \n",
    "\n",
    "We will sample from this list later to create the negative class.\n",
    "\n",
    "We define a negative gene as:\n",
    "\n",
    "- they cant be positive genes\n",
    "- they cant be ambiguos genes\n",
    "- OPTIONAL: many proteins of same class but still differt (ex. odoranr receptros)\n",
    "    - SOLUTION: solve wht pfam and interpro domains --> group proteins that have same domains --> take only one from the group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "92fc2873",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20207\n",
      "20020\n",
      "20004\n"
     ]
    }
   ],
   "source": [
    "# Start with all proteins in uniref_df\n",
    "PUTATIVE_NEGATIVE_PROTEINS = set().union(*uniref_df.protein)\n",
    "print(len(PUTATIVE_NEGATIVE_PROTEINS))\n",
    "\n",
    "# Remove positve and ambibuos\n",
    "PUTATIVE_NEGATIVE_PROTEINS = PUTATIVE_NEGATIVE_PROTEINS.difference(POSITIVE_PROTEINS)\n",
    "print(len(PUTATIVE_NEGATIVE_PROTEINS))\n",
    "PUTATIVE_NEGATIVE_PROTEINS = PUTATIVE_NEGATIVE_PROTEINS.difference(AMBIGUOS_PROTEINS)\n",
    "print(len(PUTATIVE_NEGATIVE_PROTEINS))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6a23389",
   "metadata": {},
   "source": [
    "### Optional: Remove duplicated protein domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c2f10ed5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# interpro_df = pd.read_csv(config[\"data_paths\"][\"interprot_domains\"], sep=\"\\t\")\n",
    "\n",
    "# def extract_ids(cell, prefix):\n",
    "#     if pd.isna(cell):\n",
    "#         return []\n",
    "#     # Find all matches like PFxxxxx or IPRxxxxx\n",
    "#     ids = re.findall(rf\"{prefix}\\d+\", cell)\n",
    "#     return list(set(ids))  # optional: remove duplicates\n",
    "\n",
    "# # Apply parsing\n",
    "# interpro_df[\"Pfam_list\"] = interpro_df[\"Pfam\"].apply(lambda x: extract_ids(str(x), \"PF\"))\n",
    "# interpro_df[\"InterPro_list\"] = interpro_df[\"InterPro\"].apply(lambda x: extract_ids(str(x), \"IPR\"))\n",
    "\n",
    "# interpro_df[\"n_Pfam\"] = interpro_df[\"Pfam_list\"].apply(len)\n",
    "# interpro_df[\"n_InterPro\"] = interpro_df[\"InterPro_list\"].apply(len)\n",
    "\n",
    "# interpro_df[\"all_domains\"] = interpro_df.apply(lambda row: set(row['Pfam_list']) | set(row['InterPro_list']), axis=1)\n",
    "\n",
    "# interpro_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ed9243e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Filter to current subset of prots\n",
    "# interpro_df = interpro_df[interpro_df.Entry.isin(putative_negative_protein_gene_df.protein.to_list())]\n",
    "\n",
    "# # Convert sets to frozensets for grouping (sets aren't hashable)\n",
    "# interpro_df['domains_hashable'] = interpro_df['all_domains'].apply(frozenset)\n",
    "\n",
    "# # Group by domains and aggregate\n",
    "# final_groups_df = (\n",
    "#     interpro_df[interpro_df.duplicated(subset=['domains_hashable'], keep=False)] #keep=False, mark all pccurencies\n",
    "#     .groupby('domains_hashable')['Entry']  # group by set of domains\n",
    "#     .agg(Proteins_in_Group=list, Group_Size='size')\n",
    "#         #Collect all the protein_id values in the group into a list\n",
    "#         #'size'\tCount how many items are in the group\n",
    "#     .reset_index()\n",
    "#     .sort_values('Group_Size', ascending=False)\n",
    "# )\n",
    "\n",
    "# # Convert frozensets back to sets for display\n",
    "# final_groups_df['all_domains'] = final_groups_df['domains_hashable'].apply(set)\n",
    "# final_groups_df = final_groups_df.drop(columns=['domains_hashable'])\n",
    "\n",
    "# display(final_groups_df)\n",
    "\n",
    "# # Clean up the temporary column\n",
    "# interpro_df = interpro_df.drop(columns=['domains_hashable'])\n",
    "# interpro_df = interpro_df.drop_duplicates(subset=\"all_domains\")\n",
    "\n",
    "# # Filter df\n",
    "# putative_negative_protein_gene_df = putative_negative_protein_gene_df[putative_negative_protein_gene_df.protein.isin(interpro_df.Entry.to_list())]\n",
    "\n",
    "# print(\"Remained proteins:\", len(putative_negative_protein_gene_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c59ef91b",
   "metadata": {},
   "source": [
    "### Optional: subsample to remove proteins with overlapping domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a873b45c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Flatten the Jaccard matrix (exclude diagonal if it's 1.0 self-similarity)\n",
    "# jaccard_values = jaccard_matrix[np.triu_indices_from(jaccard_matrix, k=1)]  # Upper triangle only\n",
    "# jaccard_values = jaccard_values[jaccard_values != 0]\n",
    "\n",
    "# # Create histogram\n",
    "# plt.figure(figsize=(10, 6))\n",
    "# plt.hist(jaccard_values, bins=100, edgecolor='black', alpha=0.7)\n",
    "# plt.xlabel('Jaccard Similarity', fontsize=12)\n",
    "# plt.ylabel('Frequency', fontsize=12)\n",
    "# plt.title('Distribution of Jaccard Similarities', fontsize=14)\n",
    "# plt.grid(axis='y', alpha=0.3)\n",
    "\n",
    "# # Add statistics\n",
    "# mean_val = np.mean(jaccard_values)\n",
    "# median_val = np.median(jaccard_values)\n",
    "# plt.axvline(mean_val, color='red', linestyle='--', linewidth=2, label=f'Mean: {mean_val:.3f}')\n",
    "# plt.axvline(median_val, color='green', linestyle='--', linewidth=2, label=f'Median: {median_val:.3f}')\n",
    "# plt.legend()\n",
    "\n",
    "# # Print summary statistics\n",
    "# print(\"Jaccard Similarity Statistics:\")\n",
    "# print(f\"Min: {jaccard_values.min():.4f}\")\n",
    "# print(f\"Max: {jaccard_values.max():.4f}\")\n",
    "# print(f\"Mean: {mean_val:.4f}\")\n",
    "# print(f\"Median: {median_val:.4f}\")\n",
    "# print(f\"Std: {np.std(jaccard_values):.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "09a364ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create edge list using SIMILARITY (Jaccard) as weights\n",
    "# # Leiden with ModularityVertexPartition: higher weight = stronger connection\n",
    "# threshold = median_val  # Minimum Jaccard similarity to create an edge\n",
    "\n",
    "# edges = []\n",
    "# weights = []\n",
    "# for i in range(len(jaccard_df)):\n",
    "#     for j in range(i+1, len(jaccard_df)):\n",
    "#         similarity = jaccard_matrix[i, j]\n",
    "#         if similarity > threshold:  # Use similarity directly\n",
    "#             edges.append((i, j))\n",
    "#             weights.append(similarity)  # Higher Jaccard = stronger edge\n",
    "\n",
    "# # Create igraph\n",
    "# g = ig.Graph(n=len(jaccard_df), edges=edges)\n",
    "# g.es['weight'] = weights\n",
    "\n",
    "# # Run Leiden clustering\n",
    "# partition = leidenalg.find_partition(\n",
    "#     g, \n",
    "#     leidenalg.ModularityVertexPartition,\n",
    "#     weights='weight',\n",
    "#     n_iterations=-1,  # Run until convergence\n",
    "#     seed=42 \n",
    "# )\n",
    "\n",
    "# # Add cluster labels to dataframe\n",
    "# interpro_df['Leiden_Cluster'] = partition.membership\n",
    "\n",
    "# for i, g in interpro_df.groupby(['Leiden_Cluster']):\n",
    "#     print(\"Cluster name\", i, \"Lenght:\", len(g))\n",
    "#     display(\"Memebers:\", g.Entry.unique())\n",
    "#     if i[0] == 0:\n",
    "#         break\n",
    "\n",
    "# print(f\"\\nNumber of clusters: {len(set(partition.membership))}\")\n",
    "# print(f\"Modularity: {partition.modularity:.3f}\")\n",
    "# print(f\"\\nCluster sizes:\")\n",
    "# display(interpro_df['Leiden_Cluster'].value_counts().sort_index())\n",
    "\n",
    "# # Check for singletons (clusters with only 1 member)\n",
    "# cluster_sizes = interpro_df['Leiden_Cluster'].value_counts()\n",
    "# singletons = cluster_sizes[cluster_sizes == 1]\n",
    "# print(f\"\\nSingletons (clusters of size 1): {len(singletons)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "71f95d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # retain onlyu one prot form each cluster\n",
    "# interpro_df_filtered = interpro_df.drop_duplicates(subset=[\"Leiden_Cluster\"]).copy()\n",
    "# display(interpro_df_filtered)\n",
    "\n",
    "# # Filter df\n",
    "# putative_negative_protein_gene_df = putative_negative_protein_gene_df[putative_negative_protein_gene_df.protein.isin(interpro_df_filtered.Entry.to_list())]\n",
    "# display(putative_negative_protein_gene_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c588bc01",
   "metadata": {},
   "source": [
    "## Prepare col where to sample negatives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "efbc675a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>n_proteins_positive</th>\n",
       "      <th>logits</th>\n",
       "      <th>probs</th>\n",
       "      <th>putative_negative_proteins</th>\n",
       "      <th>putative_negative_genes</th>\n",
       "      <th>n_putative_negative_proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>UniRef50_Q15233</td>\n",
       "      <td>[Q15233, Q8WXF1]</td>\n",
       "      <td>[NONO, PSPC1]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[Q15233, Q8WXF1]</td>\n",
       "      <td>[NONO, PSPC1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[1, 1]</td>\n",
       "      <td>[0.5, 0.5]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13463</th>\n",
       "      <td>UniRef50_O60828</td>\n",
       "      <td>[O60828]</td>\n",
       "      <td>[PQBP1]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[O60828]</td>\n",
       "      <td>[PQBP1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16395</th>\n",
       "      <td>UniRef50_Q5HYJ3</td>\n",
       "      <td>[Q5HYJ3]</td>\n",
       "      <td>[FAM76B]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q5HYJ3]</td>\n",
       "      <td>[FAM76B]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15000</th>\n",
       "      <td>UniRef50_Q14966</td>\n",
       "      <td>[Q14966]</td>\n",
       "      <td>[ZNF638]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q14966]</td>\n",
       "      <td>[ZNF638]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14416</th>\n",
       "      <td>UniRef50_Q2TAY7</td>\n",
       "      <td>[Q2TAY7]</td>\n",
       "      <td>[SMU1]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q2TAY7]</td>\n",
       "      <td>[SMU1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster           protein           gene  n_proteins  n_genes  \\\n",
       "404    UniRef50_Q15233  [Q15233, Q8WXF1]  [NONO, PSPC1]           2        2   \n",
       "13463  UniRef50_O60828          [O60828]        [PQBP1]           1        1   \n",
       "16395  UniRef50_Q5HYJ3          [Q5HYJ3]       [FAM76B]           1        1   \n",
       "15000  UniRef50_Q14966          [Q14966]       [ZNF638]           1        1   \n",
       "14416  UniRef50_Q2TAY7          [Q2TAY7]         [SMU1]           1        1   \n",
       "\n",
       "      proteins_positive genes_positive cluster_label  n_genes_positive  \\\n",
       "404    [Q15233, Q8WXF1]  [NONO, PSPC1]      positive                 2   \n",
       "13463          [O60828]        [PQBP1]      positive                 1   \n",
       "16395          [Q5HYJ3]       [FAM76B]      positive                 1   \n",
       "15000          [Q14966]       [ZNF638]      positive                 1   \n",
       "14416          [Q2TAY7]         [SMU1]      positive                 1   \n",
       "\n",
       "       n_proteins_positive  logits       probs putative_negative_proteins  \\\n",
       "404                      2  [1, 1]  [0.5, 0.5]                         []   \n",
       "13463                    1     [1]       [1.0]                         []   \n",
       "16395                    1     [1]       [1.0]                         []   \n",
       "15000                    1     [1]       [1.0]                         []   \n",
       "14416                    1     [1]       [1.0]                         []   \n",
       "\n",
       "      putative_negative_genes  n_putative_negative_proteins  \n",
       "404                        []                             0  \n",
       "13463                      []                             0  \n",
       "16395                      []                             0  \n",
       "15000                      []                             0  \n",
       "14416                      []                             0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create list of putative negative genes/protein        \n",
    "    # ATTENTION; putatove = neither postive neither ambigous\n",
    "uniref_df[\"putative_negative_proteins\"], uniref_df[\"putative_negative_genes\"] = zip(*uniref_df.apply(\n",
    "    lambda row: (\n",
    "        [p for p, g in zip(row[\"protein\"], row[\"gene\"]) if p in PUTATIVE_NEGATIVE_PROTEINS],\n",
    "        [g for p, g in zip(row[\"protein\"], row[\"gene\"]) if p in PUTATIVE_NEGATIVE_PROTEINS],\n",
    "    ),\n",
    "    axis=1\n",
    "))\n",
    "# count\n",
    "uniref_df[\"n_putative_negative_proteins\"] = uniref_df.putative_negative_proteins.apply(len)\n",
    "\n",
    "# ATTENTION: same NEGATIVE clusters coudl have 0 proteins\n",
    "# necause all proteins are postive OR ambigous so nothing left to sample as negative)\n",
    "\n",
    "display(uniref_df.head(5))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "331267b8",
   "metadata": {},
   "source": [
    "# *) Are there enriched clusters?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b112ced0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fraction_positive\n",
      "0.0    18826\n",
      "0.5        4\n",
      "1.0      182\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjoAAAGdCAYAAAAbudkLAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAAIlhJREFUeJzt3XtwVPX9//HXJmEDOEkAU3KRQBRBG8BkzIWiMIKmIjBUYFppUQiUoo5LyxjRhtKSTlFB0AzWrqZaITK1QmmROgVRiTgo4gCB4CUKIgFCSQIUzSahJCE5vz8c9vuL4ZaTvSSffT5m9o89e7J576fYPOfsObsOy7IsAQAAGCgs2AMAAAD4C6EDAACMRegAAABjEToAAMBYhA4AADAWoQMAAIxF6AAAAGMROgAAwFgRwR4g2FpaWnT8+HFFRUXJ4XAEexwAAHAFLMtSbW2tEhMTFRZ28eM2IR86x48fV1JSUrDHAAAANlRUVKhfv34XfTzkQycqKkrStwsVHR0d5GkAAMCV8Hg8SkpK8v4dv5iQD53zb1dFR0cTOgAAdDGXO+2Ek5EBAICxCB0AAGAsQgcAABiL0AEAAMYidAAAgLEIHQAAYCxCBwAAGIvQAQAAxiJ0AACAsQgdAABgLEIHAAAYK2RDx+12KyUlRZmZmcEeBQAA+InDsiwr2EMEk8fjUUxMjGpqavhSTwAAuogr/fsd8t9e7k/JeRsvu8/hpRMCMAkAAKEpZN+6AgAA5iN0AACAsQgdAABgLEIHAAAYi9ABAADGInQAAICxCB0AAGAsQgcAABiL0AEAAMYidAAAgLEIHQAAYCxCBwAAGIvQAQAAxiJ0AACAsQgdAABgLEIHAAAYi9ABAADGInQAAICxCB0AAGAsY0LnzJkzGjBggObPnx/sUQAAQCdhTOg88cQT+sEPfhDsMQAAQCdiROh8+eWX+uKLLzRu3LhgjwIAADqRoIfOtm3bNHHiRCUmJsrhcGjDhg1t9nG73UpOTlb37t01fPhw7dy5s9Xj8+fP15IlSwI0MQAA6CqCHjr19fVKTU2V2+2+4ONr165Vbm6u8vPztWfPHqWmpmrs2LE6ceKEJOlf//qXBg8erMGDBwdybAAA0AVEBHuAcePGXfItp4KCAs2ZM0ezZs2SJBUWFmrjxo1auXKl8vLy9NFHH2nNmjVat26d6urq1NTUpOjoaC1atOiCz9fQ0KCGhgbvfY/H49sXBAAAOo2gH9G5lMbGRpWUlCg7O9u7LSwsTNnZ2dqxY4ckacmSJaqoqNDhw4f19NNPa86cOReNnPP7x8TEeG9JSUl+fx0AACA4OnXonDp1Ss3NzYqLi2u1PS4uTlVVVbaec8GCBaqpqfHeKioqfDEqAADohIL+1pUvzZw587L7REZGKjIy0v/DAACAoOvUR3RiY2MVHh6u6urqVturq6sVHx8fpKkAAEBX0alDx+l0Kj09XcXFxd5tLS0tKi4u1ogRIzr03G63WykpKcrMzOzomAAAoJMK+ltXdXV1OnjwoPd+eXm5SktL1adPH/Xv31+5ubnKyclRRkaGsrKytGLFCtXX13uvwrLL5XLJ5XLJ4/EoJiamoy8DAAB0QkEPnd27d2vMmDHe+7m5uZKknJwcFRUVaerUqTp58qQWLVqkqqoqpaWlafPmzW1OUAYAAPguh2VZVrCHCKbzR3RqamoUHR3t0+dOztt42X0OL53g098JAEAouNK/3536HB1/4hwdAADMF7Kh43K5VFZWpl27dgV7FAAA4CchGzoAAMB8hA4AADAWoQMAAIxF6AAAAGOFbOhw1RUAAOYL2dDhqisAAMwXsqEDAADMR+gAAABjEToAAMBYhA4AADBWyIYOV10BAGC+kA0drroCAMB8IRs6AADAfIQOAAAwFqEDAACMRegAAABjEToAAMBYIRs6XF4OAID5QjZ0uLwcAADzhWzoAAAA8xE6AADAWIQOAAAwFqEDAACMRegAAABjEToAAMBYhA4AADBWyIYOHxgIAID5QjZ0+MBAAADMF7KhAwAAzEfoAAAAYxE6AADAWIQOAAAwFqEDAACMRegAAABjEToAAMBYhA4AADAWoQMAAIwVsqHDV0AAAGC+kA0dvgICAADzhWzoAAAA8xE6AADAWIQOAAAwFqEDAACMRegAAABjEToAAMBYhA4AADAWoQMAAIxF6AAAAGMROgAAwFiEDgAAMBahAwAAjEXoAAAAYxE6AADAWIQOAAAwVsiGjtvtVkpKijIzM4M9CgAA8JOQDR2Xy6WysjLt2rUr2KMAAAA/CdnQAQAA5iN0AACAsQgdAABgLEIHAAAYi9ABAADGInQAAICxCB0AAGAsQgcAABiL0AEAAMYidAAAgLEIHQAAYCxCBwAAGIvQAQAAxiJ0AACAsQgdAABgLEIHAAAYi9ABAADGInQAAICxCB0AAGCsLh8633zzjTIyMpSWlqahQ4fqpZdeCvZIAACgk4gI9gAdFRUVpW3btqlnz56qr6/X0KFDNWXKFF199dXBHg0AAARZlz+iEx4erp49e0qSGhoaZFmWLMsK8lQAAKAzCHrobNu2TRMnTlRiYqIcDoc2bNjQZh+3263k5GR1795dw4cP186dO1s9/s033yg1NVX9+vXTo48+qtjY2ABNDwAAOrOgh059fb1SU1Pldrsv+PjatWuVm5ur/Px87dmzR6mpqRo7dqxOnDjh3adXr17at2+fysvL9be//U3V1dWBGh8AAHRiQQ+dcePG6fHHH9fkyZMv+HhBQYHmzJmjWbNmKSUlRYWFherZs6dWrlzZZt+4uDilpqbq/fffv+jva2hokMfjaXUDAABmCnroXEpjY6NKSkqUnZ3t3RYWFqbs7Gzt2LFDklRdXa3a2lpJUk1NjbZt26Ybbrjhos+5ZMkSxcTEeG9JSUn+fREAACBoOnXonDp1Ss3NzYqLi2u1PS4uTlVVVZKkI0eOaNSoUUpNTdWoUaP0y1/+UsOGDbvocy5YsEA1NTXeW0VFhV9fAwAACJ4uf3l5VlaWSktLr3j/yMhIRUZG+m8gAADQaXTqIzqxsbEKDw9vc3JxdXW14uPjgzQVAADoKjp16DidTqWnp6u4uNi7raWlRcXFxRoxYkSHntvtdislJUWZmZkdHRMAAHRSQX/rqq6uTgcPHvTeLy8vV2lpqfr06aP+/fsrNzdXOTk5ysjIUFZWllasWKH6+nrNmjWrQ7/X5XLJ5XLJ4/EoJiamoy8DAAB0QkEPnd27d2vMmDHe+7m5uZKknJwcFRUVaerUqTp58qQWLVqkqqoqpaWlafPmzW1OUAYAAPguhxXi35dw/ohOTU2NoqOjffrcyXkbL7vP4aUTfPo7AQAIBVf697tTn6MDAADQESEbOpyMDACA+UI2dFwul8rKyrRr165gjwIAAPwkZEMHAACYj9ABAADGInQAAICxQjZ0OBkZAADzhWzocDIyAADmC9nQAQAA5iN0AACAsQgdAABgLFuhc+jQIV/PAQAA4HO2Quf666/XmDFj9Ne//lVnz5719UwBwVVXAACYz1bo7NmzRzfddJNyc3MVHx+vBx54QDt37vT1bH7FVVcAAJjPVuikpaXp2Wef1fHjx7Vy5UpVVlZq5MiRGjp0qAoKCnTy5ElfzwkAANBuHToZOSIiQlOmTNG6dev01FNP6eDBg5o/f76SkpI0Y8YMVVZW+mpOAACAdutQ6OzevVsPPfSQEhISVFBQoPnz5+urr77SO++8o+PHj+vuu+/21ZwAAADtFmHnhwoKCrRq1Srt379f48eP1+rVqzV+/HiFhX3bTddee62KioqUnJzsy1kBAADaxVbovPDCC/r5z3+umTNnKiEh4YL79O3bVy+//HKHhgMAAOgIW6Hz5ZdfXnYfp9OpnJwcO08PAADgE7bO0Vm1apXWrVvXZvu6dev0yiuvdHioQOBzdAAAMJ+t0FmyZIliY2PbbO/bt6+efPLJDg8VCHyODgAA5rMVOkePHtW1117bZvuAAQN09OjRDg8FAADgC7ZCp2/fvvr444/bbN+3b5+uvvrqDg8FAADgC7ZC52c/+5l+9atfaevWrWpublZzc7PeffddzZs3Tz/96U99PSMAAIAttq66Wrx4sQ4fPqw77rhDERHfPkVLS4tmzJjRZc7RAQAA5rMVOk6nU2vXrtXixYu1b98+9ejRQ8OGDdOAAQN8PR8AAIBttkLnvMGDB2vw4MG+mgUAAMCnbIVOc3OzioqKVFxcrBMnTqilpaXV4++++65PhgMAAOgIW6Ezb948FRUVacKECRo6dKgcDoev5/I7t9stt9ut5ubmYI8CAAD8xGFZltXeH4qNjfV+kWdX5/F4FBMTo5qaGkVHR/v0uZPzNl52n8NLJ/j0dwIAEAqu9O+3rcvLnU6nrr/+etvDAQAABIKt0HnkkUf07LPPysbBIAAAgICxdY7OBx98oK1bt+rNN9/UkCFD1K1bt1aPr1+/3ifDAQAAdISt0OnVq5cmT57s61kAAAB8ylborFq1ytdzAAAA+Jytc3Qk6dy5c9qyZYv+/Oc/q7a2VpJ0/Phx1dXV+Ww4AACAjrB1ROfIkSO66667dPToUTU0NOiHP/yhoqKi9NRTT6mhoUGFhYW+nhMAAKDdbB3RmTdvnjIyMvT111+rR48e3u2TJ09WcXGxz4YDAADoCFtHdN5//319+OGHcjqdrbYnJyfrP//5j08GAwAA6ChbR3RaWlou+NUJx44dU1RUVIeHCgS3262UlBRlZmYGexQAAOAntkLnzjvv1IoVK7z3HQ6H6urqlJ+f32W+FsLlcqmsrEy7du0K9igAAMBPbL119cwzz2js2LFKSUnR2bNnNW3aNH355ZeKjY3Va6+95usZAQAAbLEVOv369dO+ffu0Zs0affzxx6qrq9Ps2bN17733tjo5GQAAIJhshY4kRURE6L777vPlLAAAAD5lK3RWr159ycdnzJhhaxgAAABfshU68+bNa3W/qalJZ86ckdPpVM+ePQkdAADQKdi66urrr79udaurq9P+/fs1cuRITkYGAACdhu3vuvquQYMGaenSpW2O9gAAAASLz0JH+vYE5ePHj/vyKQEAAGyzdY7OG2+80eq+ZVmqrKzUn/70J916660+GQwAAKCjbIXOpEmTWt13OBz63ve+p9tvv13PPPOML+YCAADoMFuh09LS4us5AAAAfM6n5+gAAAB0JraO6OTm5l7xvgUFBXZ+BQAAQIfZCp29e/dq7969ampq0g033CBJOnDggMLDw3XzzTd793M4HL6ZEgAAwAZboTNx4kRFRUXplVdeUe/evSV9+yGCs2bN0qhRo/TII4/4dEh/cLvdcrvdam5uDvYoAADATxyWZVnt/aFrrrlGb7/9toYMGdJq+6effqo777yzS32WjsfjUUxMjGpqahQdHe3T507O23jZfQ4vneDT3wkAQCi40r/ftk5G9ng8OnnyZJvtJ0+eVG1trZ2nBAAA8DlboTN58mTNmjVL69ev17Fjx3Ts2DH985//1OzZszVlyhRfzwgAAGCLrXN0CgsLNX/+fE2bNk1NTU3fPlFEhGbPnq3ly5f7dEAAANA5dYVTNGyFTs+ePfX8889r+fLl+uqrryRJAwcO1FVXXeXT4QAAADqiQx8YWFlZqcrKSg0aNEhXXXWVbJzXDAAA4De2Que///2v7rjjDg0ePFjjx49XZWWlJGn27Nld4tJyAAAQGmyFzsMPP6xu3brp6NGj6tmzp3f71KlTtXnzZp8NBwAA0BG2ztF5++239dZbb6lfv36ttg8aNEhHjhzxyWAAAAAdZeuITn19fasjOeedPn1akZGRHR4KAADAF2yFzqhRo7R69WrvfYfDoZaWFi1btkxjxozx2XAAAAAdYeutq2XLlumOO+7Q7t271djYqMcee0yfffaZTp8+re3bt/t6RgAAAFtsHdEZOnSoDhw4oJEjR+ruu+9WfX29pkyZor1792rgwIG+nhEAAMCWdh/RaWpq0l133aXCwkItXLjQHzMBAAD4RLuP6HTr1k0ff/yxP2YBAADwKVtvXd133316+eWXfT0LAACAT9k6GfncuXNauXKltmzZovT09DbfcVVQUOCT4QAAADqiXaFz6NAhJScn69NPP9XNN98sSTpw4ECrfRwOh++mAwAA6IB2hc6gQYNUWVmprVu3Svr2Kx/++Mc/Ki4uzi/DAQAAdES7ztH57reTv/nmm6qvr/fpQAAAAL5i62Tk874bPgAAAJ1Ju0LH4XC0OQcn2OfkVFRUaPTo0UpJSdFNN92kdevWBXUeAADQebTrHB3LsjRz5kzvF3eePXtWDz74YJurrtavX++7CS8jIiJCK1asUFpamqqqqpSenq7x48e3mQkAAISedoVOTk5Oq/v33XefT4exIyEhQQkJCZKk+Ph4xcbG6vTp04QOAABoX+isWrXK5wNs27ZNy5cvV0lJiSorK/X6669r0qRJrfZxu91avny5qqqqlJqaqueee05ZWVltnqukpETNzc1KSkry+ZwAAKDr6dDJyL5QX1+v1NRUud3uCz6+du1a5ebmKj8/X3v27FFqaqrGjh2rEydOtNrv9OnTmjFjhl588cVAjA0AALoAW5+M7Evjxo3TuHHjLvp4QUGB5syZo1mzZkmSCgsLtXHjRq1cuVJ5eXmSpIaGBk2aNEl5eXm65ZZbLvn7Ghoa1NDQ4L3v8Xh88CoAAEBnFPQjOpfS2NiokpISZWdne7eFhYUpOztbO3bskPR/J0jffvvtmj59+mWfc8mSJYqJifHeeJsLAABzderQOXXqlJqbm9t88nJcXJyqqqokSdu3b9fatWu1YcMGpaWlKS0tTZ988slFn3PBggWqqanx3ioqKvz6GgAAQPAE/a2rjho5cqRaWlqueP/IyEjv5fEAAMBsnfqITmxsrMLDw1VdXd1qe3V1teLj44M0FQAA6Co6deg4nU6lp6eruLjYu62lpUXFxcUaMWJEh57b7XYrJSVFmZmZHR0TAAB0UkF/66qurk4HDx703i8vL1dpaan69Omj/v37Kzc3Vzk5OcrIyFBWVpZWrFih+vp671VYdrlcLrlcLnk8HsXExHT0ZQAAgE4o6KGze/dujRkzxns/NzdX0refwlxUVKSpU6fq5MmTWrRokaqqqpSWlqbNmze3OUEZAADgu4IeOqNHj77st6DPnTtXc+fODdBEAADAFJ36HB1/4hwdAADMF7Kh43K5VFZWpl27dgV7FAAA4CchGzoAAMB8hA4AADAWoQMAAIxF6AAAAGOFbOhw1RUAAOYL2dDhqisAAMwXsqEDAADMR+gAAABjEToAAMBYhA4AADBWyIYOV10BAGC+kA0drroCAMB8IRs6AADAfIQOAAAwFqEDAACMRegAAABjEToAAMBYIRs6XF4OAID5QjZ0uLwcAADzhWzoAAAA8xE6AADAWIQOAAAwFqEDAACMRegAAABjEToAAMBYhA4AADBWyIYOHxgIAID5QjZ0+MBAAADMF7KhAwAAzEfoAAAAYxE6AADAWIQOAAAwFqEDAACMRegAAABjEToAAMBYhA4AADAWoQMAAIwVsqHDV0AAAGC+kA0dvgICAADzhWzoAAAA8xE6AADAWIQOAAAwFqEDAACMRegAAABjEToAAMBYhA4AADAWoQMAAIxF6AAAAGMROgAAwFiEDgAAMBahAwAAjEXoAAAAYxE6AADAWIQOAAAwVsiGjtvtVkpKijIzM4M9CgAA8JOQDR2Xy6WysjLt2rUr2KMAAAA/CdnQAQAA5iN0AACAsQgdAABgLEIHAAAYi9ABAADGInQAAICxCB0AAGAsQgcAABiL0AEAAMYidAAAgLEIHQAAYCxCBwAAGIvQAQAAxiJ0AACAsQgdAABgLEIHAAAYi9ABAADGInQAAICxCB0AAGAsI0Jn8uTJ6t27t3784x8HexQAANCJGBE68+bN0+rVq4M9BgAA6GSMCJ3Ro0crKioq2GMAAIBOJuihs23bNk2cOFGJiYlyOBzasGFDm33cbreSk5PVvXt3DR8+XDt37gz8oAAAoMsJeujU19crNTVVbrf7go+vXbtWubm5ys/P1549e5SamqqxY8fqxIkTAZ4UAAB0NRHBHmDcuHEaN27cRR8vKCjQnDlzNGvWLElSYWGhNm7cqJUrVyovL6/dv6+hoUENDQ3e+x6Pp/1DAwCALiHoR3QupbGxUSUlJcrOzvZuCwsLU3Z2tnbs2GHrOZcsWaKYmBjvLSkpyVfjAgCATqZTh86pU6fU3NysuLi4Vtvj4uJUVVXlvZ+dna2f/OQn2rRpk/r163fJCFqwYIFqamq8t4qKCr/NDwAAgivob135wpYtW65438jISEVGRvpxGgAA0Fl06iM6sbGxCg8PV3V1davt1dXVio+PD9JUAACgq+jUoeN0OpWenq7i4mLvtpaWFhUXF2vEiBEdem63262UlBRlZmZ2dEwAANBJBf2tq7q6Oh08eNB7v7y8XKWlperTp4/69++v3Nxc5eTkKCMjQ1lZWVqxYoXq6+u9V2HZ5XK55HK55PF4FBMT09GXAQAAOqGgh87u3bs1ZswY7/3c3FxJUk5OjoqKijR16lSdPHlSixYtUlVVldLS0rR58+Y2JygDAAB8V9BDZ/To0bIs65L7zJ07V3Pnzg3QRAAAwBSd+hwdAACAjgjZ0OFkZAAAzBeyoeNyuVRWVqZdu3YFexQAAOAnIRs6AADAfIQOAAAwFqEDAACMFbKhw8nIAACYL2RDh5ORAQAwX8iGDgAAMB+hAwAAjEXoAAAAYxE6AADAWCEbOlx1BQCA+UI2dLjqCgAA84Vs6AAAAPMROgAAwFiEDgAAMBahAwAAjEXoAAAAYxE6AADAWCEbOnyODgAA5gvZ0OFzdAAAMF/Ihg4AADAfoQMAAIxF6AAAAGMROgAAwFiEDgAAMBahAwAAjBUR7AGCxe12y+12q7m5OdijAPCj5LyNl93n8NIJAZgEQDCE7BEdPkcHAADzhWzoAAAA8xE6AADAWIQOAAAwFqEDAACMRegAAABjEToAAMBYhA4AADAWoQMAAIxF6AAAAGMROgAAwFghGzput1spKSnKzMwM9igAAMBPQjZ0+K4rAADMF7KhAwAAzEfoAAAAY0UEe4BgsyxLkuTxeHz+3C0NZy67jz9+L4D/w3+HgP8E87+v8897/u/4xTisy+1huGPHjikpKSnYYwAAABsqKirUr1+/iz4e8qHT0tKi48ePKyoqSg6Hw2fP6/F4lJSUpIqKCkVHR/vsedEa6xw4rHVgsM6BwToHhj/X2bIs1dbWKjExUWFhFz8TJ+TfugoLC7tkCXZUdHQ0/xEFAOscOKx1YLDOgcE6B4a/1jkmJuay+3AyMgAAMBahAwAAjEXo+ElkZKTy8/MVGRkZ7FGMxjoHDmsdGKxzYLDOgdEZ1jnkT0YGAADm4ogOAAAwFqEDAACMRegAAABjEToAAMBYhE4HuN1uJScnq3v37ho+fLh27tx5yf3XrVunG2+8Ud27d9ewYcO0adOmAE3atbVnnV966SWNGjVKvXv3Vu/evZWdnX3Z/13wrfb+ez5vzZo1cjgcmjRpkn8HNEh71/qbb76Ry+VSQkKCIiMjNXjwYP7/4wq0d51XrFihG264QT169FBSUpIefvhhnT17NkDTdk3btm3TxIkTlZiYKIfDoQ0bNlz2Z9577z3dfPPNioyM1PXXX6+ioiL/DmnBljVr1lhOp9NauXKl9dlnn1lz5syxevXqZVVXV19w/+3bt1vh4eHWsmXLrLKyMuu3v/2t1a1bN+uTTz4J8ORdS3vXedq0aZbb7bb27t1rff7559bMmTOtmJgY69ixYwGevGtp7zqfV15ebl1zzTXWqFGjrLvvvjsww3Zx7V3rhoYGKyMjwxo/frz1wQcfWOXl5dZ7771nlZaWBnjyrqW96/zqq69akZGR1quvvmqVl5dbb731lpWQkGA9/PDDAZ68a9m0aZO1cOFCa/369ZYk6/XXX7/k/ocOHbJ69uxp5ebmWmVlZdZzzz1nhYeHW5s3b/bbjISOTVlZWZbL5fLeb25uthITE60lS5ZccP977rnHmjBhQqttw4cPtx544AG/ztnVtXedv+vcuXNWVFSU9corr/hrRCPYWedz585Zt9xyi/WXv/zFysnJIXSuUHvX+oUXXrCuu+46q7GxMVAjGqG96+xyuazbb7+91bbc3Fzr1ltv9eucJrmS0HnsscesIUOGtNo2depUa+zYsX6bi7eubGhsbFRJSYmys7O928LCwpSdna0dO3Zc8Gd27NjRan9JGjt27EX3h711/q4zZ86oqalJffr08deYXZ7ddf7DH/6gvn37avbs2YEY0wh21vqNN97QiBEj5HK5FBcXp6FDh+rJJ59Uc3NzoMbucuys8y233KKSkhLv21uHDh3Spk2bNH78+IDMHCqC8bcw5L/U045Tp06publZcXFxrbbHxcXpiy++uODPVFVVXXD/qqoqv83Z1dlZ5+/69a9/rcTExDb/YeH/2FnnDz74QC+//LJKS0sDMKE57Kz1oUOH9O677+ree+/Vpk2bdPDgQT300ENqampSfn5+IMbucuys87Rp03Tq1CmNHDlSlmXp3LlzevDBB/Wb3/wmECOHjIv9LfR4PPrf//6nHj16+Px3ckQHxlq6dKnWrFmj119/Xd27dw/2OMaora3V9OnT9dJLLyk2NjbY4xivpaVFffv21Ysvvqj09HRNnTpVCxcuVGFhYbBHM8p7772nJ598Us8//7z27Nmj9evXa+PGjVq8eHGwR0MHcUTHhtjYWIWHh6u6urrV9urqasXHx1/wZ+Lj49u1P+yt83lPP/20li5dqi1btuimm27y55hdXnvX+auvvtLhw4c1ceJE77aWlhZJUkREhPbv36+BAwf6d+guys6/6YSEBHXr1k3h4eHebd///vdVVVWlxsZGOZ1Ov87cFdlZ59/97neaPn26fvGLX0iShg0bpvr6et1///1auHChwsI4LuALF/tbGB0d7ZejORJHdGxxOp1KT09XcXGxd1tLS4uKi4s1YsSIC/7MiBEjWu0vSe+8885F94e9dZakZcuWafHixdq8ebMyMjICMWqX1t51vvHGG/XJJ5+otLTUe/vRj36kMWPGqLS0VElJSYEcv0ux82/61ltv1cGDB70xKUkHDhxQQkICkXMRdtb5zJkzbWLmfFxafCWkzwTlb6HfTnM23Jo1a6zIyEirqKjIKisrs+6//36rV69eVlVVlWVZljV9+nQrLy/Pu//27dutiIgI6+mnn7Y+//xzKz8/n8vLr0B713np0qWW0+m0/vGPf1iVlZXeW21tbbBeQpfQ3nX+Lq66unLtXeujR49aUVFR1ty5c639+/db//73v62+fftajz/+eLBeQpfQ3nXOz8+3oqKirNdee806dOiQ9fbbb1sDBw607rnnnmC9hC6htrbW2rt3r7V3715LklVQUGDt3bvXOnLkiGVZlpWXl2dNnz7du//5y8sfffRR6/PPP7fcbjeXl3dmzz33nNW/f3/L6XRaWVlZ1kcffeR97LbbbrNycnJa7f/3v//dGjx4sOV0Oq0hQ4ZYGzduDPDEXVN71nnAgAGWpDa3/Pz8wA/exbT33/P/j9Bpn/au9YcffmgNHz7cioyMtK677jrriSeesM6dOxfgqbue9qxzU1OT9fvf/94aOHCg1b17dyspKcl66KGHrK+//jrwg3chW7duveD/555f25ycHOu2225r8zNpaWmW0+m0rrvuOmvVqlV+ndFhWRyTAwAAZuIcHQAAYCxCBwAAGIvQAQAAxiJ0AACAsQgdAABgLEIHAAAYi9ABAADGInQAAICxCB0AAGAsQgcAABiL0AEAAMYidAAAgLH+H3Hva7r9MPJyAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# FRACTION OF CLUSTER OCCUPIED BY POSITIVE GENES\n",
    "uniref_df[\"fraction_positive\"] = uniref_df.n_genes_positive / uniref_df.n_genes\n",
    "uniref_df.sort_values(by=\"fraction_positive\", inplace=True, ascending=False)\n",
    "\n",
    "# hsit\n",
    "uniref_df.fraction_positive.plot(kind=\"hist\", logy=True, bins=50)\n",
    "print(uniref_df.fraction_positive.value_counts().sort_index())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c1e47b0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique total genes: 20207\n",
      "Unique total positive genes: 187\n",
      "\n",
      "Significant enriched gene sets: 0 / 19012\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>enrichment_prob</th>\n",
       "      <th>enrichment_pvalue</th>\n",
       "      <th>enrichment_pvalue_adj</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>UniRef50_Q15233</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14249</th>\n",
       "      <td>UniRef50_O43395</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15381</th>\n",
       "      <td>UniRef50_Q15459</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8364</th>\n",
       "      <td>UniRef50_Q99459</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18927</th>\n",
       "      <td>UniRef50_Q00839</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7220</th>\n",
       "      <td>UniRef50_Q9UQ35</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1309</th>\n",
       "      <td>UniRef50_Q8NHY2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17140</th>\n",
       "      <td>UniRef50_P57052</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17717</th>\n",
       "      <td>UniRef50_P48729</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6897</th>\n",
       "      <td>UniRef50_Q9UKV3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.009254</td>\n",
       "      <td>0.96671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster  n_genes  n_genes_positive  enrichment_prob  \\\n",
       "404    UniRef50_Q15233        2                 2         0.000085   \n",
       "14249  UniRef50_O43395        1                 1         0.009254   \n",
       "15381  UniRef50_Q15459        1                 1         0.009254   \n",
       "8364   UniRef50_Q99459        1                 1         0.009254   \n",
       "18927  UniRef50_Q00839        1                 1         0.009254   \n",
       "7220   UniRef50_Q9UQ35        1                 1         0.009254   \n",
       "1309   UniRef50_Q8NHY2        1                 1         0.009254   \n",
       "17140  UniRef50_P57052        1                 1         0.009254   \n",
       "17717  UniRef50_P48729        1                 1         0.009254   \n",
       "6897   UniRef50_Q9UKV3        1                 1         0.009254   \n",
       "\n",
       "       enrichment_pvalue  enrichment_pvalue_adj  significant  \n",
       "404             0.000085                0.96671        False  \n",
       "14249           0.009254                0.96671        False  \n",
       "15381           0.009254                0.96671        False  \n",
       "8364            0.009254                0.96671        False  \n",
       "18927           0.009254                0.96671        False  \n",
       "7220            0.009254                0.96671        False  \n",
       "1309            0.009254                0.96671        False  \n",
       "17140           0.009254                0.96671        False  \n",
       "17717           0.009254                0.96671        False  \n",
       "6897            0.009254                0.96671        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# HYPERGEMOETRIC TEST\n",
    "\n",
    "# Filter genesets with enough genes\n",
    "    # ATTENTION\n",
    "    # CANNOT USE TOO SMALL DRAWING (n) --> test power comes from the changing probabilities over several draws, known as sampling without replacemen\n",
    "uniref_df_big = uniref_df[uniref_df.n_genes >= 1].copy()\n",
    "\n",
    "# Define background\n",
    "total_genes = len(set().union(*uniref_df_big[\"gene\"]))\n",
    "total_positive = len(set().union(*uniref_df_big[\"genes_positive\"]))\n",
    "print(f\"Unique total genes: {total_genes}\")\n",
    "print(f\"Unique total positive genes: {total_positive}\")\n",
    "\n",
    "# Run enrichment test\n",
    "results = uniref_df_big.apply(\n",
    "    lambda row: pd.Series(\n",
    "        dataf.per_cluster_hypergeom_test(\n",
    "            row[\"n_genes\"], \n",
    "            row[\"n_genes_positive\"], \n",
    "            total_genes, \n",
    "            total_positive\n",
    "        )\n",
    "    ), \n",
    "    axis=1\n",
    ")\n",
    "results.columns = [\"enrichment_prob\", \"enrichment_pvalue\"]\n",
    "\n",
    "# Combine results\n",
    "df_uniref_enriched = pd.concat([uniref_df_big, results], axis=1)\n",
    "df_uniref_enriched = df_uniref_enriched.sort_values(\"enrichment_pvalue\")\n",
    "\n",
    "# Multiple testing correction (FDR)\n",
    "df_uniref_enriched[\"enrichment_pvalue_adj\"] = multipletests(\n",
    "    df_uniref_enriched[\"enrichment_pvalue\"], \n",
    "    method=\"fdr_bh\"\n",
    ")[1]\n",
    "df_uniref_enriched[\"significant\"] = df_uniref_enriched[\"enrichment_pvalue_adj\"] < 0.05\n",
    "\n",
    "# Summary\n",
    "print(f\"\\nSignificant enriched gene sets: {df_uniref_enriched['significant'].sum()} / {len(df_uniref_enriched)}\")\n",
    "display(\n",
    "    df_uniref_enriched[\n",
    "        [\"UniRef50_Cluster\", \"n_genes\", \"n_genes_positive\", \n",
    "         \"enrichment_prob\", \"enrichment_pvalue\", \n",
    "         \"enrichment_pvalue_adj\", \"significant\"]\n",
    "    ].head(10)\n",
    ")\n",
    "\n",
    "# only enriched\n",
    "df_uniref_significant = df_uniref_enriched[df_uniref_enriched['significant'] == True]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fe0fc7e",
   "metadata": {},
   "source": [
    "# Create Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d84ae75",
   "metadata": {},
   "source": [
    "## Positive class sampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "280ce5ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>n_proteins_positive</th>\n",
       "      <th>logits</th>\n",
       "      <th>probs</th>\n",
       "      <th>putative_negative_proteins</th>\n",
       "      <th>putative_negative_genes</th>\n",
       "      <th>n_putative_negative_proteins</th>\n",
       "      <th>fraction_positive</th>\n",
       "      <th>proteins_sampled</th>\n",
       "      <th>genes_sampled</th>\n",
       "      <th>n_proteins_sampled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14249</th>\n",
       "      <td>UniRef50_O43395</td>\n",
       "      <td>[O43395]</td>\n",
       "      <td>[PRPF3]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[O43395]</td>\n",
       "      <td>[PRPF3]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[O43395]</td>\n",
       "      <td>[PRPF3]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8364</th>\n",
       "      <td>UniRef50_Q99459</td>\n",
       "      <td>[Q99459]</td>\n",
       "      <td>[CDC5L]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q99459]</td>\n",
       "      <td>[CDC5L]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Q99459]</td>\n",
       "      <td>[CDC5L]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster   protein     gene  n_proteins  n_genes  \\\n",
       "14249  UniRef50_O43395  [O43395]  [PRPF3]           1        1   \n",
       "8364   UniRef50_Q99459  [Q99459]  [CDC5L]           1        1   \n",
       "\n",
       "      proteins_positive genes_positive cluster_label  n_genes_positive  \\\n",
       "14249          [O43395]        [PRPF3]      positive                 1   \n",
       "8364           [Q99459]        [CDC5L]      positive                 1   \n",
       "\n",
       "       n_proteins_positive logits  probs putative_negative_proteins  \\\n",
       "14249                    1    [1]  [1.0]                         []   \n",
       "8364                     1    [1]  [1.0]                         []   \n",
       "\n",
       "      putative_negative_genes  n_putative_negative_proteins  \\\n",
       "14249                      []                             0   \n",
       "8364                       []                             0   \n",
       "\n",
       "       fraction_positive proteins_sampled genes_sampled  n_proteins_sampled  \n",
       "14249                1.0         [O43395]       [PRPF3]                   1  \n",
       "8364                 1.0         [Q99459]       [CDC5L]                   1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(186, 19)\n"
     ]
    }
   ],
   "source": [
    "# Filter only positive clusters\n",
    "uniref_df_pos = uniref_df[uniref_df.cluster_label == \"positive\"].copy()\n",
    "\n",
    "# sample or take all postive proteins?\n",
    "if not config[\"dataset_creation\"][\"sample_positive_genes\"]:\n",
    "    SAMPLE_N_POSITIVE = uniref_df_pos.n_proteins_positive.max() # the number to sample is the max possbile\n",
    "else:\n",
    "    SAMPLE_N_POSITIVE = config[\"dataset_creation\"][\"number_positive_proteins_to_sample_per_cluster\"]\n",
    "\n",
    "uniref_df_pos[\"proteins_sampled\"], uniref_df_pos[\"genes_sampled\"] = zip(\n",
    "    *uniref_df_pos.apply(\n",
    "            dataf.sample_sampled_from_single_row, \n",
    "            axis=1, \n",
    "            min_sample_n=SAMPLE_N_POSITIVE, \n",
    "            gene_col=\"genes_positive\",  # ATTENTION_: sample from correct columns --> the psotive gene here\n",
    "            prot_col=\"proteins_positive\", \n",
    "            probs_col=\"probs\")\n",
    ")\n",
    "uniref_df_pos[\"n_proteins_sampled\"] = uniref_df_pos.proteins_sampled.apply(len)\n",
    "\n",
    "display(uniref_df_pos.head(2))\n",
    "print(uniref_df_pos.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7688c2f1",
   "metadata": {},
   "source": [
    "## Negative Sampling\n",
    "\n",
    "- do not use clusters that are negative but have 0 negative prots (ex all are ambiguos)\n",
    "- do not take neagtive proteins that come form psotive lcusters (even if the single proteins are negative)\n",
    "- sample proteins wiht uniform distr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e013631c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>n_proteins_positive</th>\n",
       "      <th>logits</th>\n",
       "      <th>probs</th>\n",
       "      <th>putative_negative_proteins</th>\n",
       "      <th>putative_negative_genes</th>\n",
       "      <th>n_putative_negative_proteins</th>\n",
       "      <th>fraction_positive</th>\n",
       "      <th>proteins_sampled</th>\n",
       "      <th>genes_sampled</th>\n",
       "      <th>n_proteins_sampled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>UniRef50_P18669</td>\n",
       "      <td>[P18669, P15259, Q8N0Y7]</td>\n",
       "      <td>[PGAM1, PGAM2, PGAM4]</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[P18669, P15259, Q8N0Y7]</td>\n",
       "      <td>[PGAM1, PGAM2, PGAM4]</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[P18669, P15259, Q8N0Y7]</td>\n",
       "      <td>[PGAM1, PGAM2, PGAM4]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>UniRef50_Q9UDX3</td>\n",
       "      <td>[Q9UDX4, Q9UDX3, B5MCN3]</td>\n",
       "      <td>[SEC14L3, SEC14L4, SEC14L6]</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q9UDX4, Q9UDX3, B5MCN3]</td>\n",
       "      <td>[SEC14L3, SEC14L4, SEC14L6]</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Q9UDX4, Q9UDX3, B5MCN3]</td>\n",
       "      <td>[SEC14L3, SEC14L4, SEC14L6]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>UniRef50_A6NM11</td>\n",
       "      <td>[A6NMS7, A6NM11, O60309]</td>\n",
       "      <td>[LRRC37A, LRRC37A2, LRRC37A3]</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A6NMS7, A6NM11, O60309]</td>\n",
       "      <td>[LRRC37A, LRRC37A2, LRRC37A3]</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[A6NMS7, A6NM11, O60309]</td>\n",
       "      <td>[LRRC37A, LRRC37A2, LRRC37A3]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>UniRef50_P17024</td>\n",
       "      <td>[Q8NDP4, Q7L945, C9JN71, P17024]</td>\n",
       "      <td>[ZNF439, ZNF627, ZNF878, ZNF20]</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q8NDP4, Q7L945, C9JN71, P17024]</td>\n",
       "      <td>[ZNF439, ZNF627, ZNF878, ZNF20]</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Q8NDP4, Q7L945, C9JN71, P17024]</td>\n",
       "      <td>[ZNF439, ZNF627, ZNF878, ZNF20]</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>UniRef50_Q15155</td>\n",
       "      <td>[Q15155, Q5JPE7, P69849]</td>\n",
       "      <td>[NOMO1, NOMO2, NOMO3]</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q15155, Q5JPE7, P69849]</td>\n",
       "      <td>[NOMO1, NOMO2, NOMO3]</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Q15155, Q5JPE7, P69849]</td>\n",
       "      <td>[NOMO1, NOMO2, NOMO3]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniRef50_Cluster                           protein  \\\n",
       "115  UniRef50_P18669          [P18669, P15259, Q8N0Y7]   \n",
       "117  UniRef50_Q9UDX3          [Q9UDX4, Q9UDX3, B5MCN3]   \n",
       "118  UniRef50_A6NM11          [A6NMS7, A6NM11, O60309]   \n",
       "73   UniRef50_P17024  [Q8NDP4, Q7L945, C9JN71, P17024]   \n",
       "120  UniRef50_Q15155          [Q15155, Q5JPE7, P69849]   \n",
       "\n",
       "                                gene  n_proteins  n_genes proteins_positive  \\\n",
       "115            [PGAM1, PGAM2, PGAM4]           3        3                []   \n",
       "117      [SEC14L3, SEC14L4, SEC14L6]           3        3                []   \n",
       "118    [LRRC37A, LRRC37A2, LRRC37A3]           3        3                []   \n",
       "73   [ZNF439, ZNF627, ZNF878, ZNF20]           4        4                []   \n",
       "120            [NOMO1, NOMO2, NOMO3]           3        3                []   \n",
       "\n",
       "    genes_positive cluster_label  n_genes_positive  n_proteins_positive  \\\n",
       "115             []      negative                 0                    0   \n",
       "117             []      negative                 0                    0   \n",
       "118             []      negative                 0                    0   \n",
       "73              []      negative                 0                    0   \n",
       "120             []      negative                 0                    0   \n",
       "\n",
       "    logits probs        putative_negative_proteins  \\\n",
       "115     []    []          [P18669, P15259, Q8N0Y7]   \n",
       "117     []    []          [Q9UDX4, Q9UDX3, B5MCN3]   \n",
       "118     []    []          [A6NMS7, A6NM11, O60309]   \n",
       "73      []    []  [Q8NDP4, Q7L945, C9JN71, P17024]   \n",
       "120     []    []          [Q15155, Q5JPE7, P69849]   \n",
       "\n",
       "             putative_negative_genes  n_putative_negative_proteins  \\\n",
       "115            [PGAM1, PGAM2, PGAM4]                             3   \n",
       "117      [SEC14L3, SEC14L4, SEC14L6]                             3   \n",
       "118    [LRRC37A, LRRC37A2, LRRC37A3]                             3   \n",
       "73   [ZNF439, ZNF627, ZNF878, ZNF20]                             4   \n",
       "120            [NOMO1, NOMO2, NOMO3]                             3   \n",
       "\n",
       "     fraction_positive                  proteins_sampled  \\\n",
       "115                0.0          [P18669, P15259, Q8N0Y7]   \n",
       "117                0.0          [Q9UDX4, Q9UDX3, B5MCN3]   \n",
       "118                0.0          [A6NMS7, A6NM11, O60309]   \n",
       "73                 0.0  [Q8NDP4, Q7L945, C9JN71, P17024]   \n",
       "120                0.0          [Q15155, Q5JPE7, P69849]   \n",
       "\n",
       "                       genes_sampled  n_proteins_sampled  \n",
       "115            [PGAM1, PGAM2, PGAM4]                   3  \n",
       "117      [SEC14L3, SEC14L4, SEC14L6]                   3  \n",
       "118    [LRRC37A, LRRC37A2, LRRC37A3]                   3  \n",
       "73   [ZNF439, ZNF627, ZNF878, ZNF20]                   4  \n",
       "120            [NOMO1, NOMO2, NOMO3]                   3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18810, 19)\n"
     ]
    }
   ],
   "source": [
    "# Filter only negative clusters\n",
    "    # sample from clusters that do nto contain psotive (negative label of before)\n",
    "uniref_df_neg = uniref_df[\n",
    "    (uniref_df.n_putative_negative_proteins > 0) & # some lcuster maybe haev only ambigus genes\n",
    "    (uniref_df.cluster_label == \"negative\") # avoid proteins from proteins class are used as eagtive\n",
    "    ].copy() \n",
    "\n",
    "# Sample or nor sample?\n",
    "if config[\"dataset_creation\"][\"sample_negative_genes\"]:\n",
    "    print(\"Sampling negative class\")\n",
    "    \n",
    "    uniref_df_neg[\"proteins_sampled\"], uniref_df_neg[\"genes_sampled\"] = zip(\n",
    "        *uniref_df_neg.apply(\n",
    "                dataf.sample_sampled_from_single_row, \n",
    "                axis=1, \n",
    "                min_sample_n=SAMPLE_N_POSITIVE, \n",
    "                gene_col=\"putative_negative_genes\",  # ATTENTION\n",
    "                prot_col=\"putative_negative_proteins\", \n",
    "                probs_col=None # uniform\n",
    "                )\n",
    "    )\n",
    "else:\n",
    "    uniref_df_neg[\"proteins_sampled\"], uniref_df_neg[\"genes_sampled\"] = uniref_df_neg[\"putative_negative_proteins\"], uniref_df_neg[\"putative_negative_genes\"] \n",
    "\n",
    "uniref_df_neg[\"n_proteins_sampled\"] = uniref_df_neg.proteins_sampled.apply(len)\n",
    "\n",
    "# ATTENTION: same clusters coudl have 0 proteins to sample from (all proteins are postive OR ambigous so nothing left to sample as negative)\n",
    "uniref_df_neg = uniref_df_neg[uniref_df_neg.n_proteins_sampled > 0]\n",
    "\n",
    "display(uniref_df_neg.head())\n",
    "print(uniref_df_neg.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "652723dc",
   "metadata": {},
   "source": [
    "## Make dataset\n",
    "\n",
    "- concat pos and neg\n",
    "- expand to have\n",
    "\n",
    "| Cluster | protein name | label | seq |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7e2c2374",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>n_proteins_positive</th>\n",
       "      <th>logits</th>\n",
       "      <th>probs</th>\n",
       "      <th>putative_negative_proteins</th>\n",
       "      <th>putative_negative_genes</th>\n",
       "      <th>n_putative_negative_proteins</th>\n",
       "      <th>fraction_positive</th>\n",
       "      <th>proteins_sampled</th>\n",
       "      <th>genes_sampled</th>\n",
       "      <th>n_proteins_sampled</th>\n",
       "      <th>label_single_prot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14249</th>\n",
       "      <td>UniRef50_O43395</td>\n",
       "      <td>[O43395]</td>\n",
       "      <td>[PRPF3]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[O43395]</td>\n",
       "      <td>[PRPF3]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[O43395]</td>\n",
       "      <td>[PRPF3]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8364</th>\n",
       "      <td>UniRef50_Q99459</td>\n",
       "      <td>[Q99459]</td>\n",
       "      <td>[CDC5L]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q99459]</td>\n",
       "      <td>[CDC5L]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Q99459]</td>\n",
       "      <td>[CDC5L]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15381</th>\n",
       "      <td>UniRef50_Q15459</td>\n",
       "      <td>[Q15459]</td>\n",
       "      <td>[SF3A1]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q15459]</td>\n",
       "      <td>[SF3A1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Q15459]</td>\n",
       "      <td>[SF3A1]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17140</th>\n",
       "      <td>UniRef50_P57052</td>\n",
       "      <td>[P57052]</td>\n",
       "      <td>[RBM11]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[P57052]</td>\n",
       "      <td>[RBM11]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[P57052]</td>\n",
       "      <td>[RBM11]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19002</th>\n",
       "      <td>UniRef50_Q02040</td>\n",
       "      <td>[Q02040]</td>\n",
       "      <td>[AKAP17A]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[Q02040]</td>\n",
       "      <td>[AKAP17A]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Q02040]</td>\n",
       "      <td>[AKAP17A]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster   protein       gene  n_proteins  n_genes  \\\n",
       "14249  UniRef50_O43395  [O43395]    [PRPF3]           1        1   \n",
       "8364   UniRef50_Q99459  [Q99459]    [CDC5L]           1        1   \n",
       "15381  UniRef50_Q15459  [Q15459]    [SF3A1]           1        1   \n",
       "17140  UniRef50_P57052  [P57052]    [RBM11]           1        1   \n",
       "19002  UniRef50_Q02040  [Q02040]  [AKAP17A]           1        1   \n",
       "\n",
       "      proteins_positive genes_positive cluster_label  n_genes_positive  \\\n",
       "14249          [O43395]        [PRPF3]      positive                 1   \n",
       "8364           [Q99459]        [CDC5L]      positive                 1   \n",
       "15381          [Q15459]        [SF3A1]      positive                 1   \n",
       "17140          [P57052]        [RBM11]      positive                 1   \n",
       "19002          [Q02040]      [AKAP17A]      positive                 1   \n",
       "\n",
       "       n_proteins_positive logits  probs putative_negative_proteins  \\\n",
       "14249                    1    [1]  [1.0]                         []   \n",
       "8364                     1    [1]  [1.0]                         []   \n",
       "15381                    1    [1]  [1.0]                         []   \n",
       "17140                    1    [1]  [1.0]                         []   \n",
       "19002                    1    [1]  [1.0]                         []   \n",
       "\n",
       "      putative_negative_genes  n_putative_negative_proteins  \\\n",
       "14249                      []                             0   \n",
       "8364                       []                             0   \n",
       "15381                      []                             0   \n",
       "17140                      []                             0   \n",
       "19002                      []                             0   \n",
       "\n",
       "       fraction_positive proteins_sampled genes_sampled  n_proteins_sampled  \\\n",
       "14249                1.0         [O43395]       [PRPF3]                   1   \n",
       "8364                 1.0         [Q99459]       [CDC5L]                   1   \n",
       "15381                1.0         [Q15459]       [SF3A1]                   1   \n",
       "17140                1.0         [P57052]       [RBM11]                   1   \n",
       "19002                1.0         [Q02040]     [AKAP17A]                   1   \n",
       "\n",
       "       label_single_prot  \n",
       "14249                  1  \n",
       "8364                   1  \n",
       "15381                  1  \n",
       "17140                  1  \n",
       "19002                  1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(18996, 20)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add labels\n",
    "uniref_df_pos[\"label_single_prot\"] = 1 # all the sample dproteins in a cluster eceive the same label --> cannot have 2 differt labels in same clustr\n",
    "uniref_df_neg[\"label_single_prot\"] = 0\n",
    "\n",
    "# merge --> each entry is a cluster\n",
    "dataset_df = pd.concat([uniref_df_pos, uniref_df_neg])\n",
    "display(dataset_df.head())\n",
    "display(dataset_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d7e3a8e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATTENTION: the number of proteins sampled is less than the total ~20000 of Swissport, as we subsampled only N, so we may have lsot some proteins\n",
      "20187 / 20210\n",
      "\n",
      "Duplicated rows/proteins:  0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniRef50_O43395</td>\n",
       "      <td>O43395</td>\n",
       "      <td>PRPF3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniRef50_Q99459</td>\n",
       "      <td>Q99459</td>\n",
       "      <td>CDC5L</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniRef50_Q15459</td>\n",
       "      <td>Q15459</td>\n",
       "      <td>SF3A1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UniRef50_P57052</td>\n",
       "      <td>P57052</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UniRef50_Q02040</td>\n",
       "      <td>Q02040</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20182</th>\n",
       "      <td>UniRef50_Q03111</td>\n",
       "      <td>Q03111</td>\n",
       "      <td>MLLT1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20183</th>\n",
       "      <td>UniRef50_Q03060</td>\n",
       "      <td>Q03060</td>\n",
       "      <td>CREM</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20184</th>\n",
       "      <td>UniRef50_Q03052</td>\n",
       "      <td>Q03052</td>\n",
       "      <td>POU3F1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20185</th>\n",
       "      <td>UniRef50_Q03014</td>\n",
       "      <td>Q03014</td>\n",
       "      <td>HHEX</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20186</th>\n",
       "      <td>UniRef50_Q04759</td>\n",
       "      <td>Q04759</td>\n",
       "      <td>PRKCQ</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20187 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster protein     gene  label\n",
       "0      UniRef50_O43395  O43395    PRPF3      1\n",
       "1      UniRef50_Q99459  Q99459    CDC5L      1\n",
       "2      UniRef50_Q15459  Q15459    SF3A1      1\n",
       "3      UniRef50_P57052  P57052    RBM11      1\n",
       "4      UniRef50_Q02040  Q02040  AKAP17A      1\n",
       "...                ...     ...      ...    ...\n",
       "20182  UniRef50_Q03111  Q03111    MLLT1      0\n",
       "20183  UniRef50_Q03060  Q03060     CREM      0\n",
       "20184  UniRef50_Q03052  Q03052   POU3F1      0\n",
       "20185  UniRef50_Q03014  Q03014     HHEX      0\n",
       "20186  UniRef50_Q04759  Q04759    PRKCQ      0\n",
       "\n",
       "[20187 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create long df\n",
    "\n",
    "# smaller df\n",
    "cols_to_keep = ['UniRef50_Cluster', 'proteins_sampled',\n",
    "                'genes_sampled', 'label_single_prot']\n",
    "dataset_df_small = dataset_df[cols_to_keep].copy()\n",
    "\n",
    "# Create a new DataFrame where each row corresponds to a (protein, gene) pair\n",
    "rows = []\n",
    "for _, row in dataset_df_small.iterrows():\n",
    "    cluster = row[\"UniRef50_Cluster\"] # single vlaue\n",
    "    label = row[\"label_single_prot\"] # single value\n",
    "    proteins = row[\"proteins_sampled\"] # list\n",
    "    genes = row[\"genes_sampled\"] # list\n",
    "\n",
    "    # zip ensures 1-to-1 pairing between proteins and genes\n",
    "    for prot, gene in zip(proteins, genes):\n",
    "        rows.append({\n",
    "            \"UniRef50_Cluster\": cluster,\n",
    "            \"protein\": prot,\n",
    "            \"gene\": gene,\n",
    "            \"label\": label\n",
    "        })\n",
    "\n",
    "df_long = pd.DataFrame(rows)\n",
    "\n",
    "print(\"ATTENTION: the number of proteins sampled is less than the total ~20000 of Swissport, as we subsampled only N, so we may have lsot some proteins\")\n",
    "print(df_long.protein.nunique(), \"/\", proteome_uniprot.shape[0])\n",
    "\n",
    "# checj fro duplcates\n",
    "print(\"\\nDuplicated rows/proteins: \", df_long.shape[0] - df_long.protein.nunique())\n",
    "\n",
    "df_long"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e4781a",
   "metadata": {},
   "source": [
    "## Assure 1:N ratio in postive:negative class\n",
    "After previos sampling sampling you will have\n",
    "- | pos prot | << | neg prot | â†’ more thna 1:3 ration\n",
    "- Need to subsampled negative class\n",
    "- Subsample negative class\n",
    "    - Subsample the remaining cluster to have a final ratio 1:3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "96d4857d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "label\n",
      "0    20000\n",
      "1      187\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniRef50_O43395</td>\n",
       "      <td>O43395</td>\n",
       "      <td>PRPF3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniRef50_Q99459</td>\n",
       "      <td>Q99459</td>\n",
       "      <td>CDC5L</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniRef50_Q15459</td>\n",
       "      <td>Q15459</td>\n",
       "      <td>SF3A1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UniRef50_P57052</td>\n",
       "      <td>P57052</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UniRef50_Q02040</td>\n",
       "      <td>Q02040</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20182</th>\n",
       "      <td>UniRef50_Q03111</td>\n",
       "      <td>Q03111</td>\n",
       "      <td>MLLT1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20183</th>\n",
       "      <td>UniRef50_Q03060</td>\n",
       "      <td>Q03060</td>\n",
       "      <td>CREM</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20184</th>\n",
       "      <td>UniRef50_Q03052</td>\n",
       "      <td>Q03052</td>\n",
       "      <td>POU3F1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20185</th>\n",
       "      <td>UniRef50_Q03014</td>\n",
       "      <td>Q03014</td>\n",
       "      <td>HHEX</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20186</th>\n",
       "      <td>UniRef50_Q04759</td>\n",
       "      <td>Q04759</td>\n",
       "      <td>PRKCQ</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20187 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster protein     gene  label\n",
       "0      UniRef50_O43395  O43395    PRPF3      1\n",
       "1      UniRef50_Q99459  Q99459    CDC5L      1\n",
       "2      UniRef50_Q15459  Q15459    SF3A1      1\n",
       "3      UniRef50_P57052  P57052    RBM11      1\n",
       "4      UniRef50_Q02040  Q02040  AKAP17A      1\n",
       "...                ...     ...      ...    ...\n",
       "20182  UniRef50_Q03111  Q03111    MLLT1      0\n",
       "20183  UniRef50_Q03060  Q03060     CREM      0\n",
       "20184  UniRef50_Q03052  Q03052   POU3F1      0\n",
       "20185  UniRef50_Q03014  Q03014     HHEX      0\n",
       "20186  UniRef50_Q04759  Q04759    PRKCQ      0\n",
       "\n",
       "[20187 rows x 4 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if config[\"dataset_creation\"][\"sample_negative_genes\"]:\n",
    "    print(\"Sampling\")\n",
    "    \n",
    "    # list of smpled positve proteins\n",
    "    positive_proteins_tmp = set(df_long[df_long.label == 1][\"protein\"].unique())\n",
    "\n",
    "    # list of prtein in MIXED cluster (wiht inside btoh postive and negative prteins)\n",
    "        # ATTENTUON: check no ptortien in mixed clusters\n",
    "    proteins_in_mixed_clusters = df_long.groupby('UniRef50_Cluster').filter(lambda x: x['label'].nunique() > 1).sort_values(by=\"UniRef50_Cluster\")\n",
    "    print(\"priteins in mixed clusters:\", len(proteins_in_mixed_clusters))\n",
    "    proteins_in_mixed_clusters = set(proteins_in_mixed_clusters[\"protein\"].unique())\n",
    "\n",
    "    # proteins to NOT remove\n",
    "    proteins_not_to_remove = proteins_in_mixed_clusters.union(positive_proteins_tmp)\n",
    "\n",
    "    # give col\n",
    "    df_long[\"can_be_removed\"] = ~df_long[\"protein\"].isin(proteins_not_to_remove)\n",
    "\n",
    "\n",
    "    # decide fraction to susample\n",
    "\n",
    "    # Number of positive proteins\n",
    "    n_positive = df_long[df_long[\"label\"] == 1].shape[0]\n",
    "    # Number desidered negative\n",
    "    n_desired_negatives = n_positive * config[\"dataset_creation\"][\"mult_factor_negative_class\"]\n",
    "    # Removable (negative) proteins\n",
    "    removable_negatives = df_long[(df_long[\"label\"] == 0) & (df_long[\"can_be_removed\"])]\n",
    "    # Determine fraction to subsample\n",
    "    frac_to_sample = n_desired_negatives / len(removable_negatives)\n",
    "    frac_to_sample = min(frac_to_sample, 1.0)  # cannot sample more than available\n",
    "\n",
    "    frac_to_sample\n",
    "\n",
    "\n",
    "    # susample\n",
    "    removable = df_long[df_long[\"can_be_removed\"]]\n",
    "    keep = df_long[~df_long[\"can_be_removed\"]]\n",
    "\n",
    "    subsampled_removable = removable.sample(frac=frac_to_sample)\n",
    "\n",
    "    df_subsampled = pd.concat([keep, subsampled_removable]).reset_index(drop=True) # Combine back with the protected proteins\n",
    "\n",
    "else:\n",
    "    df_subsampled = df_long.copy()\n",
    "\n",
    "print(df_subsampled.label.value_counts())\n",
    "df_subsampled"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0f9feee",
   "metadata": {},
   "source": [
    "## Give Sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c2ab40d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>label</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniRef50_O43395</td>\n",
       "      <td>O43395</td>\n",
       "      <td>PRPF3</td>\n",
       "      <td>1</td>\n",
       "      <td>MALSKRELDELKPWIEKTVKRVLGFSEPTVVTAALNCVGKGMDKKK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniRef50_Q99459</td>\n",
       "      <td>Q99459</td>\n",
       "      <td>CDC5L</td>\n",
       "      <td>1</td>\n",
       "      <td>MPRIMIKGGVWRNTEDEILKAAVMKYGKNQWSRIASLLHRKSAKQC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniRef50_Q15459</td>\n",
       "      <td>Q15459</td>\n",
       "      <td>SF3A1</td>\n",
       "      <td>1</td>\n",
       "      <td>MPAGPVQAVPPPPPVPTEPKQPTEEEASSKEDSAPSKPVVGIIYPP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UniRef50_P57052</td>\n",
       "      <td>P57052</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>1</td>\n",
       "      <td>MFPAQEEADRTVFVGNLEARVREEILYELFLQAGPLTKVTICKDRE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UniRef50_Q02040</td>\n",
       "      <td>Q02040</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>1</td>\n",
       "      <td>MAAATIVHDTSEAVELCPAYGLYLKPITKMTISVALPQLKQPGKSI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20182</th>\n",
       "      <td>UniRef50_Q03111</td>\n",
       "      <td>Q03111</td>\n",
       "      <td>MLLT1</td>\n",
       "      <td>0</td>\n",
       "      <td>MDNQCTVQVRLELGHRAQLRKKPTTEGFTHDWMVFVRGPEQCDIQH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20183</th>\n",
       "      <td>UniRef50_Q03060</td>\n",
       "      <td>Q03060</td>\n",
       "      <td>CREM</td>\n",
       "      <td>0</td>\n",
       "      <td>MTMETVESQHDGSITASLTESKSAHVQTQTGQNSIPALAQVSVAGS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20184</th>\n",
       "      <td>UniRef50_Q03052</td>\n",
       "      <td>Q03052</td>\n",
       "      <td>POU3F1</td>\n",
       "      <td>0</td>\n",
       "      <td>MATTAQYLPRGPGGGAGGTGPLMHPDAAAAAAAAAAAERLHAGAAY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20185</th>\n",
       "      <td>UniRef50_Q03014</td>\n",
       "      <td>Q03014</td>\n",
       "      <td>HHEX</td>\n",
       "      <td>0</td>\n",
       "      <td>MQYPHPGPAAGAVGVPLYAPTPLLQPAHPTPFYIEDILGRGPAAPT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20186</th>\n",
       "      <td>UniRef50_Q04759</td>\n",
       "      <td>Q04759</td>\n",
       "      <td>PRKCQ</td>\n",
       "      <td>0</td>\n",
       "      <td>MSPFLRIGLSNFDCGSCQSCQGEAVNPYCAVLVKEYVESENGQMYI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20187 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster protein     gene  label  \\\n",
       "0      UniRef50_O43395  O43395    PRPF3      1   \n",
       "1      UniRef50_Q99459  Q99459    CDC5L      1   \n",
       "2      UniRef50_Q15459  Q15459    SF3A1      1   \n",
       "3      UniRef50_P57052  P57052    RBM11      1   \n",
       "4      UniRef50_Q02040  Q02040  AKAP17A      1   \n",
       "...                ...     ...      ...    ...   \n",
       "20182  UniRef50_Q03111  Q03111    MLLT1      0   \n",
       "20183  UniRef50_Q03060  Q03060     CREM      0   \n",
       "20184  UniRef50_Q03052  Q03052   POU3F1      0   \n",
       "20185  UniRef50_Q03014  Q03014     HHEX      0   \n",
       "20186  UniRef50_Q04759  Q04759    PRKCQ      0   \n",
       "\n",
       "                                                sequence  \n",
       "0      MALSKRELDELKPWIEKTVKRVLGFSEPTVVTAALNCVGKGMDKKK...  \n",
       "1      MPRIMIKGGVWRNTEDEILKAAVMKYGKNQWSRIASLLHRKSAKQC...  \n",
       "2      MPAGPVQAVPPPPPVPTEPKQPTEEEASSKEDSAPSKPVVGIIYPP...  \n",
       "3      MFPAQEEADRTVFVGNLEARVREEILYELFLQAGPLTKVTICKDRE...  \n",
       "4      MAAATIVHDTSEAVELCPAYGLYLKPITKMTISVALPQLKQPGKSI...  \n",
       "...                                                  ...  \n",
       "20182  MDNQCTVQVRLELGHRAQLRKKPTTEGFTHDWMVFVRGPEQCDIQH...  \n",
       "20183  MTMETVESQHDGSITASLTESKSAHVQTQTGQNSIPALAQVSVAGS...  \n",
       "20184  MATTAQYLPRGPGGGAGGTGPLMHPDAAAAAAAAAAAERLHAGAAY...  \n",
       "20185  MQYPHPGPAAGAVGVPLYAPTPLLQPAHPTPFYIEDILGRGPAAPT...  \n",
       "20186  MSPFLRIGLSNFDCGSCQSCQGEAVNPYCAVLVKEYVESENGQMYI...  \n",
       "\n",
       "[20187 rows x 5 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### create df to join\n",
    "join_df_tmp = proteome_uniprot[[\"Entry\", \"Sequence\"]].copy().rename(columns={\"Sequence\": \"sequence\"})\n",
    "\n",
    "# join\n",
    "    # ATTNETION: we shoudl NOT lose any col\n",
    "long_df_seq = pd.merge(how=\"inner\", left=df_subsampled, right=join_df_tmp, left_on=\"protein\", right_on=\"Entry\")\n",
    "\n",
    "# drp cols\n",
    "long_df_seq.drop(columns=[\"Entry\"], inplace=True)\n",
    "\n",
    "long_df_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26f7864e",
   "metadata": {},
   "source": [
    "## Stratify dataset\n",
    "\n",
    "We can stratify based on many metrics:\n",
    "1. labels\n",
    "2. PREVALENT metric\n",
    "3. **based on cluster**\n",
    "    - Put proteins coming from same cluster in different sets or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "31e628c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "label      0    1\n",
      "set              \n",
      "test    4005   36\n",
      "train  15995  150\n",
      "val        0    1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>label</th>\n",
       "      <th>sequence</th>\n",
       "      <th>set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniRef50_O43395</td>\n",
       "      <td>O43395</td>\n",
       "      <td>PRPF3</td>\n",
       "      <td>1</td>\n",
       "      <td>MALSKRELDELKPWIEKTVKRVLGFSEPTVVTAALNCVGKGMDKKK...</td>\n",
       "      <td>val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniRef50_Q99459</td>\n",
       "      <td>Q99459</td>\n",
       "      <td>CDC5L</td>\n",
       "      <td>1</td>\n",
       "      <td>MPRIMIKGGVWRNTEDEILKAAVMKYGKNQWSRIASLLHRKSAKQC...</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniRef50_Q15459</td>\n",
       "      <td>Q15459</td>\n",
       "      <td>SF3A1</td>\n",
       "      <td>1</td>\n",
       "      <td>MPAGPVQAVPPPPPVPTEPKQPTEEEASSKEDSAPSKPVVGIIYPP...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UniRef50_P57052</td>\n",
       "      <td>P57052</td>\n",
       "      <td>RBM11</td>\n",
       "      <td>1</td>\n",
       "      <td>MFPAQEEADRTVFVGNLEARVREEILYELFLQAGPLTKVTICKDRE...</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UniRef50_Q02040</td>\n",
       "      <td>Q02040</td>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>1</td>\n",
       "      <td>MAAATIVHDTSEAVELCPAYGLYLKPITKMTISVALPQLKQPGKSI...</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20182</th>\n",
       "      <td>UniRef50_Q03111</td>\n",
       "      <td>Q03111</td>\n",
       "      <td>MLLT1</td>\n",
       "      <td>0</td>\n",
       "      <td>MDNQCTVQVRLELGHRAQLRKKPTTEGFTHDWMVFVRGPEQCDIQH...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20183</th>\n",
       "      <td>UniRef50_Q03060</td>\n",
       "      <td>Q03060</td>\n",
       "      <td>CREM</td>\n",
       "      <td>0</td>\n",
       "      <td>MTMETVESQHDGSITASLTESKSAHVQTQTGQNSIPALAQVSVAGS...</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20184</th>\n",
       "      <td>UniRef50_Q03052</td>\n",
       "      <td>Q03052</td>\n",
       "      <td>POU3F1</td>\n",
       "      <td>0</td>\n",
       "      <td>MATTAQYLPRGPGGGAGGTGPLMHPDAAAAAAAAAAAERLHAGAAY...</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20185</th>\n",
       "      <td>UniRef50_Q03014</td>\n",
       "      <td>Q03014</td>\n",
       "      <td>HHEX</td>\n",
       "      <td>0</td>\n",
       "      <td>MQYPHPGPAAGAVGVPLYAPTPLLQPAHPTPFYIEDILGRGPAAPT...</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20186</th>\n",
       "      <td>UniRef50_Q04759</td>\n",
       "      <td>Q04759</td>\n",
       "      <td>PRKCQ</td>\n",
       "      <td>0</td>\n",
       "      <td>MSPFLRIGLSNFDCGSCQSCQGEAVNPYCAVLVKEYVESENGQMYI...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20187 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniRef50_Cluster protein     gene  label  \\\n",
       "0      UniRef50_O43395  O43395    PRPF3      1   \n",
       "1      UniRef50_Q99459  Q99459    CDC5L      1   \n",
       "2      UniRef50_Q15459  Q15459    SF3A1      1   \n",
       "3      UniRef50_P57052  P57052    RBM11      1   \n",
       "4      UniRef50_Q02040  Q02040  AKAP17A      1   \n",
       "...                ...     ...      ...    ...   \n",
       "20182  UniRef50_Q03111  Q03111    MLLT1      0   \n",
       "20183  UniRef50_Q03060  Q03060     CREM      0   \n",
       "20184  UniRef50_Q03052  Q03052   POU3F1      0   \n",
       "20185  UniRef50_Q03014  Q03014     HHEX      0   \n",
       "20186  UniRef50_Q04759  Q04759    PRKCQ      0   \n",
       "\n",
       "                                                sequence    set  \n",
       "0      MALSKRELDELKPWIEKTVKRVLGFSEPTVVTAALNCVGKGMDKKK...    val  \n",
       "1      MPRIMIKGGVWRNTEDEILKAAVMKYGKNQWSRIASLLHRKSAKQC...  train  \n",
       "2      MPAGPVQAVPPPPPVPTEPKQPTEEEASSKEDSAPSKPVVGIIYPP...   test  \n",
       "3      MFPAQEEADRTVFVGNLEARVREEILYELFLQAGPLTKVTICKDRE...  train  \n",
       "4      MAAATIVHDTSEAVELCPAYGLYLKPITKMTISVALPQLKQPGKSI...  train  \n",
       "...                                                  ...    ...  \n",
       "20182  MDNQCTVQVRLELGHRAQLRKKPTTEGFTHDWMVFVRGPEQCDIQH...   test  \n",
       "20183  MTMETVESQHDGSITASLTESKSAHVQTQTGQNSIPALAQVSVAGS...  train  \n",
       "20184  MATTAQYLPRGPGGGAGGTGPLMHPDAAAAAAAAAAAERLHAGAAY...  train  \n",
       "20185  MQYPHPGPAAGAVGVPLYAPTPLLQPAHPTPFYIEDILGRGPAAPT...  train  \n",
       "20186  MSPFLRIGLSNFDCGSCQSCQGEAVNPYCAVLVKEYVESENGQMYI...   test  \n",
       "\n",
       "[20187 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "import pandas as pd\n",
    "\n",
    "# --- group-level stratified split ---\n",
    "\n",
    "# pslit by cluster and then ive back set-label to each row\n",
    "\n",
    "# One representative row per UniRef50_Cluster\n",
    "cluster_df = long_df_seq.groupby('UniRef50_Cluster').first().reset_index()\n",
    "\n",
    "# Stratified split by label, but at the cluster level\n",
    "    # ists of unique cluster IDs.\n",
    "train_clusters, temp_clusters = train_test_split(\n",
    "    cluster_df['UniRef50_Cluster'],\n",
    "    test_size=0.2,\n",
    "    stratify=cluster_df['label'],\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# Split the temporary set equally into validation and test\n",
    "temp_labels = cluster_df.loc[cluster_df['UniRef50_Cluster'].isin(temp_clusters), 'label']\n",
    "val_clusters, test_clusters = train_test_split(\n",
    "    temp_clusters,\n",
    "    test_size=0.5,\n",
    "    stratify=temp_labels,\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# --- assign back to the full dataframe ---\n",
    "#Every row in long_df_seq inherits its set label based on its UniRef50_Cluster\n",
    "long_df_seq['set'] = ''\n",
    "long_df_seq.loc[long_df_seq['UniRef50_Cluster'].isin(train_clusters), 'set'] = 'train'\n",
    "long_df_seq.loc[long_df_seq['UniRef50_Cluster'].isin(test_clusters), 'set'] = 'test'\n",
    "\n",
    "# ATTENTION: ingore val\n",
    "long_df_seq.loc[long_df_seq['UniRef50_Cluster'].isin(val_clusters), 'set'] = 'val'\n",
    "long_df_seq.loc[long_df_seq['set'] == \"val\", 'set'] = 'test'\n",
    "long_df_seq.loc[long_df_seq.loc[long_df_seq['set'] == 'test'].index[0], 'set'] = 'val' \n",
    "\n",
    "# Check split distribution\n",
    "print(long_df_seq.groupby('set')['label'].value_counts().unstack(fill_value=0))\n",
    "display(long_df_seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8093d783",
   "metadata": {},
   "source": [
    "# Checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "528f3d26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n",
      "{'P49761', 'Q9NQ29', 'Q3ZCQ8', 'P02545', 'Q9H2U1', 'P27695', 'Q9UKZ4', 'Q12972', 'Q8IZP6', 'Q6K0P9', 'Q8IZD2', 'P0DOY5', 'P0DPI4', 'Q96SB4', 'Q96FV9', 'Q9GZY0', 'Q6PJT7', 'P49760', 'P19544', 'O15527', 'P0DPR3', 'Q8TDX7', 'Q86XP3'}\n",
      "{'Q9NQ29', 'Q8IZP6', 'P0DPR3', 'Q8TDX7', 'P0DOY5', 'P0DPI4', 'Q9GZY0'}\n",
      "{'P49761', 'Q3ZCQ8', 'P02545', 'Q9H2U1', 'P27695', 'Q9UKZ4', 'Q12972', 'Q6K0P9', 'Q8IZD2', 'P0DOY5', 'Q96FV9', 'P0DPI4', 'Q96SB4', 'Q6PJT7', 'P49760', 'P19544', 'O15527', 'P0DPR3', 'Q86XP3'}\n",
      "{'P0DPR3', 'P0DPI4', 'P0DOY5'}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Series([], Name: count, dtype: int64)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n",
      "True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniRef50_Cluster</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>proteins_positive</th>\n",
       "      <th>genes_positive</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>n_genes_positive</th>\n",
       "      <th>n_proteins_positive</th>\n",
       "      <th>logits</th>\n",
       "      <th>probs</th>\n",
       "      <th>putative_negative_proteins</th>\n",
       "      <th>putative_negative_genes</th>\n",
       "      <th>n_putative_negative_proteins</th>\n",
       "      <th>fraction_positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>UniRef50_Q9Y383</td>\n",
       "      <td>[Q9Y383, Q9NQ29]</td>\n",
       "      <td>[LUC7L2, LUC7L]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[Q9Y383]</td>\n",
       "      <td>[LUC7L2]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[Q9NQ29]</td>\n",
       "      <td>[LUC7L]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634</th>\n",
       "      <td>UniRef50_Q9UBU9</td>\n",
       "      <td>[Q9UBU9, Q9GZY0]</td>\n",
       "      <td>[NXF1, NXF2]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[Q9UBU9]</td>\n",
       "      <td>[NXF1]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[Q9GZY0]</td>\n",
       "      <td>[NXF2]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>UniRef50_Q9HC98</td>\n",
       "      <td>[Q9HC98, Q8TDX7]</td>\n",
       "      <td>[NEK6, NEK7]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[Q9HC98]</td>\n",
       "      <td>[NEK6]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[Q8TDX7]</td>\n",
       "      <td>[NEK7]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>UniRef50_O15541</td>\n",
       "      <td>[O15541, Q8IZP6]</td>\n",
       "      <td>[RNF113A, RNF113B]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[O15541]</td>\n",
       "      <td>[RNF113A]</td>\n",
       "      <td>positive</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>[Q8IZP6]</td>\n",
       "      <td>[RNF113B]</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniRef50_Cluster           protein                gene  n_proteins  \\\n",
       "229  UniRef50_Q9Y383  [Q9Y383, Q9NQ29]     [LUC7L2, LUC7L]           2   \n",
       "634  UniRef50_Q9UBU9  [Q9UBU9, Q9GZY0]        [NXF1, NXF2]           2   \n",
       "430  UniRef50_Q9HC98  [Q9HC98, Q8TDX7]        [NEK6, NEK7]           2   \n",
       "495  UniRef50_O15541  [O15541, Q8IZP6]  [RNF113A, RNF113B]           2   \n",
       "\n",
       "     n_genes proteins_positive genes_positive cluster_label  n_genes_positive  \\\n",
       "229        2          [Q9Y383]       [LUC7L2]      positive                 1   \n",
       "634        2          [Q9UBU9]         [NXF1]      positive                 1   \n",
       "430        2          [Q9HC98]         [NEK6]      positive                 1   \n",
       "495        2          [O15541]      [RNF113A]      positive                 1   \n",
       "\n",
       "     n_proteins_positive logits  probs putative_negative_proteins  \\\n",
       "229                    1    [1]  [1.0]                   [Q9NQ29]   \n",
       "634                    1    [1]  [1.0]                   [Q9GZY0]   \n",
       "430                    1    [1]  [1.0]                   [Q8TDX7]   \n",
       "495                    1    [1]  [1.0]                   [Q8IZP6]   \n",
       "\n",
       "    putative_negative_genes  n_putative_negative_proteins  fraction_positive  \n",
       "229                 [LUC7L]                             1                0.5  \n",
       "634                  [NXF2]                             1                0.5  \n",
       "430                  [NEK7]                             1                0.5  \n",
       "495               [RNF113B]                             1                0.5  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All intial psotive genes are presnet?\n",
    "missing_pos_genes = INITIAL_POSITIVE_GENES - set(uniref_df.gene.explode())\n",
    "print(missing_pos_genes)\n",
    "\n",
    "# All uniprot protein have been used?\n",
    "    # most probably they are \n",
    "        # 1. ambigus proteins\n",
    "        # 2. neagtive protein in positive vluster, so they cant have been sampled\n",
    "        # 3. not present inn any uniref cluster\n",
    "unused_proteins = set(proteome_uniprot.Entry).difference(long_df_seq.protein)\n",
    "unused_minus_ambiguos = unused_proteins.difference(AMBIGUOS_PROTEINS)\n",
    "prot_neg_in_pos_cluster = set(uniref_df.loc[uniref_df.protein.apply(lambda x: bool(set(x) & unused_proteins)), 'putative_negative_proteins'].explode())\n",
    "unsued_minus_prot_neg_in_pos_cluster = unused_proteins.difference(prot_neg_in_pos_cluster)\n",
    "all_uniref = set(uniref_df.protein.explode())\n",
    "unsued_minus_all_uniref = unused_proteins.difference(all_uniref)\n",
    "#uniref_df[uniref_df.protein.apply(lambda x: len(set(x).intersection(unused_proteins)) > 0)]\n",
    "\n",
    "print(unused_proteins)\n",
    "print(unused_minus_ambiguos)\n",
    "print(unsued_minus_prot_neg_in_pos_cluster)\n",
    "print(unsued_minus_all_uniref)\n",
    "\n",
    "# duplicated proteins?\n",
    "display(long_df_seq.protein.value_counts().loc[lambda x: x > 1])\n",
    "\n",
    "# Have we really removed ambogus proteins\n",
    "print(AMBIGUOS_PROTEINS.intersection(long_df_seq.protein))\n",
    "print(\"P13727\" in set(long_df_seq.protein))\n",
    "\n",
    "\n",
    "uniref_df[uniref_df.protein.apply(lambda x: len(set(x).intersection(prot_neg_in_pos_cluster)) > 0)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beff5a28",
   "metadata": {},
   "source": [
    "# Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1de8069a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/ESMSec/data/cell_location/nucleus-speckle_Uniprot-locations_03_dataset_final_sample-neg=False.csv\n"
     ]
    }
   ],
   "source": [
    "save_path = os.path.join(\n",
    "    os.path.dirname(positive_genes_path),\n",
    "    f\"{dataset_name}_03_dataset_final_sample-neg={config['dataset_creation']['sample_negative_genes']}.csv\"\n",
    ")\n",
    "long_df_seq.to_csv(save_path, index=False)\n",
    "\n",
    "print(save_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esm2_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
